{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "e4b5017d",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import shutil\n",
    "import json\n",
    "import zlib\n",
    "import traceback\n",
    "import urllib.request\n",
    "\n",
    "import pandas as pd\n",
    "from Bio import Entrez\n",
    "from Bio import SearchIO\n",
    "from Bio.SearchIO.HmmerIO.hmmer3_text import Hmmer3TextParser\n",
    "\n",
    "from eaglib.alignment import SeqsProfileInfo, SeqsProfile"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "fa006294",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Found existing installation: EAGLE 0.0.1\n",
      "Uninstalling EAGLE-0.0.1:\n",
      "  Successfully uninstalled EAGLE-0.0.1\n",
      "Processing /media/olga/Data/Denis/EAGLE/dist/EAGLE-0.0.1-py3-none-any.whl\n",
      "Requirement already satisfied: DendroPy>=4.4.0 in /home/olga/miniconda3/envs/eagle/lib/python3.8/site-packages (from EAGLE==0.0.1) (4.5.2)\n",
      "Requirement already satisfied: Deprecated>=1.2.9 in /home/olga/miniconda3/envs/eagle/lib/python3.8/site-packages (from EAGLE==0.0.1) (1.2.13)\n",
      "Requirement already satisfied: numpy>=1.14.3 in /home/olga/miniconda3/envs/eagle/lib/python3.8/site-packages (from EAGLE==0.0.1) (1.23.0)\n",
      "Requirement already satisfied: pandas>=0.22.0 in /home/olga/miniconda3/envs/eagle/lib/python3.8/site-packages (from EAGLE==0.0.1) (1.4.3)\n",
      "Requirement already satisfied: matplotlib>=2.2.3 in /home/olga/miniconda3/envs/eagle/lib/python3.8/site-packages (from EAGLE==0.0.1) (3.5.2)\n",
      "Requirement already satisfied: biopython>=1.72 in /home/olga/miniconda3/envs/eagle/lib/python3.8/site-packages (from EAGLE==0.0.1) (1.79)\n",
      "Requirement already satisfied: scipy>=1.1.0 in /home/olga/miniconda3/envs/eagle/lib/python3.8/site-packages (from EAGLE==0.0.1) (1.8.1)\n",
      "Requirement already satisfied: networkx>=2.5 in /home/olga/miniconda3/envs/eagle/lib/python3.8/site-packages (from EAGLE==0.0.1) (2.8.4)\n",
      "Requirement already satisfied: jsondler>=0.0.4 in /home/olga/miniconda3/envs/eagle/lib/python3.8/site-packages (from EAGLE==0.0.1) (0.0.4)\n",
      "Requirement already satisfied: pyaml>=3.12 in /home/olga/miniconda3/envs/eagle/lib/python3.8/site-packages (from EAGLE==0.0.1) (21.10.1)\n",
      "Requirement already satisfied: psutil>=5.6.1 in /home/olga/miniconda3/envs/eagle/lib/python3.8/site-packages (from EAGLE==0.0.1) (5.9.1)\n",
      "Requirement already satisfied: setuptools in /home/olga/miniconda3/envs/eagle/lib/python3.8/site-packages (from DendroPy>=4.4.0->EAGLE==0.0.1) (62.6.0)\n",
      "Requirement already satisfied: wrapt<2,>=1.10 in /home/olga/miniconda3/envs/eagle/lib/python3.8/site-packages (from Deprecated>=1.2.9->EAGLE==0.0.1) (1.14.1)\n",
      "Requirement already satisfied: packaging>=20.0 in /home/olga/miniconda3/envs/eagle/lib/python3.8/site-packages (from matplotlib>=2.2.3->EAGLE==0.0.1) (21.3)\n",
      "Requirement already satisfied: pillow>=6.2.0 in /home/olga/miniconda3/envs/eagle/lib/python3.8/site-packages (from matplotlib>=2.2.3->EAGLE==0.0.1) (9.2.0)\n",
      "Requirement already satisfied: fonttools>=4.22.0 in /home/olga/miniconda3/envs/eagle/lib/python3.8/site-packages (from matplotlib>=2.2.3->EAGLE==0.0.1) (4.33.3)\n",
      "Requirement already satisfied: kiwisolver>=1.0.1 in /home/olga/miniconda3/envs/eagle/lib/python3.8/site-packages (from matplotlib>=2.2.3->EAGLE==0.0.1) (1.4.3)\n",
      "Requirement already satisfied: cycler>=0.10 in /home/olga/miniconda3/envs/eagle/lib/python3.8/site-packages (from matplotlib>=2.2.3->EAGLE==0.0.1) (0.11.0)\n",
      "Requirement already satisfied: python-dateutil>=2.7 in /home/olga/miniconda3/envs/eagle/lib/python3.8/site-packages (from matplotlib>=2.2.3->EAGLE==0.0.1) (2.8.2)\n",
      "Requirement already satisfied: pyparsing>=2.2.1 in /home/olga/miniconda3/envs/eagle/lib/python3.8/site-packages (from matplotlib>=2.2.3->EAGLE==0.0.1) (3.0.9)\n",
      "Requirement already satisfied: pytz>=2020.1 in /home/olga/miniconda3/envs/eagle/lib/python3.8/site-packages (from pandas>=0.22.0->EAGLE==0.0.1) (2022.1)\n",
      "Requirement already satisfied: PyYAML in /home/olga/miniconda3/envs/eagle/lib/python3.8/site-packages (from pyaml>=3.12->EAGLE==0.0.1) (6.0)\n",
      "Requirement already satisfied: six>=1.5 in /home/olga/miniconda3/envs/eagle/lib/python3.8/site-packages (from python-dateutil>=2.7->matplotlib>=2.2.3->EAGLE==0.0.1) (1.16.0)\n",
      "Installing collected packages: EAGLE\n",
      "Successfully installed EAGLE-0.0.1\n"
     ]
    }
   ],
   "source": [
    "! pip uninstall -y EAGLE\n",
    "! pip install ../../dist/EAGLE-0.0.1-py3-none-any.whl"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "664ecbd8",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_taxonomy(tax_id):\n",
    "    tax_keys = [\"superkingdom\", \"phylum\", \"clade\", \"class\", \"order\", \"family\", \"genus\", \"species\"]\n",
    "    tax_dict = {tax_key: None for tax_key in tax_keys}\n",
    "    \n",
    "    record = Entrez.efetch(db=\"taxonomy\", id=tax_id, retmode='xml')\n",
    "    tax_info = Entrez.read(record)[0]\n",
    "    tax_dict[\"species\"] = tax_info['ScientificName']\n",
    "    \n",
    "    for lin_tax in tax_info['LineageEx']:\n",
    "        if lin_tax['Rank'] in tax_dict:\n",
    "            tax_dict[lin_tax['Rank']] = lin_tax['ScientificName']\n",
    "            \n",
    "    return [tax_dict[tax_key] for tax_key in tax_keys]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "6f52d58b",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "GCF_000762265.1 Methanobacterium formicicum strain=BRM9\n",
      "GCF_001458655.1 Methanobacterium formicicum\n",
      "GCF_002813085.1 Methanobrevibacter smithii strain=KB11\n",
      "GCF_022846155.1 Methanobrevibacter smithii strain=CE91-St67\n",
      "GCF_022846175.1 Methanobrevibacter smithii strain=CE91-St68\n",
      "GCF_001889405.1 Methanohalophilus halophilus strain=Z-7982\n",
      "GCF_019263745.1 Methanospirillum hungatei strain=GP1\n",
      "GCF_010706455.1 Methanosarcina mazei strain=zm-15\n",
      "GCF_019669945.1 Methanosarcina mazei strain=TMA\n",
      "GCF_004799605.1 Halobacterium salinarum strain=91-R6\n",
      "GCF_009729015.1 Acidianus ambivalens strain=LEI 10\n",
      "GCF_002215405.1 Sulfolobus acidocaldarius strain=Y14 18-5\n",
      "GCF_002215445.1 Sulfolobus acidocaldarius strain=Y14 20-20\n",
      "GCF_002215485.1 Sulfolobus acidocaldarius strain=Y14 16-22\n",
      "GCF_002215525.1 Sulfolobus acidocaldarius strain=Y14 13-1\n",
      "GCF_002215565.1 Sulfolobus acidocaldarius strain=DG1\n",
      "GCF_019175305.1 Saccharolobus shibatae strain=S38A\n",
      "GCF_019175325.1 Saccharolobus shibatae strain=BEU9\n",
      "GCF_000968355.2 Saccharolobus solfataricus strain=SARC-B\n",
      "GCF_000968395.2 Saccharolobus solfataricus strain=SARC-C\n",
      "GCF_000968435.2 Saccharolobus solfataricus strain=SULA\n",
      "GCF_003852095.1 Saccharolobus solfataricus strain=SARC-O\n",
      "GCF_003852115.1 Saccharolobus solfataricus strain=SUL120\n",
      "GCF_003852135.1 Saccharolobus solfataricus strain=SULG\n",
      "GCF_003852155.1 Saccharolobus solfataricus strain=SULM\n",
      "GCF_003852175.1 Saccharolobus solfataricus strain=SARC-N\n",
      "GCF_003852195.1 Saccharolobus solfataricus strain=SARC-H\n",
      "GCF_003852215.1 Saccharolobus solfataricus strain=SARC-I\n",
      "GCF_900079115.1 Saccharolobus solfataricus strain=P1\n",
      "GCF_902384015.1 Methanosphaera stadtmanae\n",
      "GCF_002355655.1 Halorubrum trapanicum strain=CBA1232\n",
      "GCF_018228765.1 Halorubrum sodomense strain=MBLA0099\n",
      "GCF_001190965.1 Haloferax gibbonsii strain=ARA6\n",
      "GCF_014969745.1 Haloferax gibbonsii strain=LR2-5\n",
      "GCF_002945325.1 Methanococcus maripaludis strain=DSM 2067\n",
      "GCF_003052125.1 Methanococcus maripaludis strain=OS7mut1\n",
      "GCF_019669925.1 Methanobrevibacter arboriphilus strain=SA\n",
      "GCF_003201835.2 Acidianus brierleyi strain=DSM 1651\n",
      "GCF_009729035.1 Stygiolobus azoricus strain=FC6\n",
      "GCF_001266655.1 Metallosphaera sedula strain=ARS50-1\n",
      "GCF_001266675.1 Metallosphaera sedula strain=ARS50-2\n",
      "GCF_001266695.1 Metallosphaera sedula strain=ARS120-1\n",
      "GCF_001266715.1 Metallosphaera sedula strain=ARS120-2\n",
      "GCF_001266735.1 Metallosphaera sedula strain=SARC-M1\n",
      "GCF_005222525.1 Metallosphaera prunae strain=Ron 12/II\n",
      "GCF_002214585.1 Thermococcus profundus strain=DT 5432\n",
      "GCF_001592435.1 Thermococcus peptonophilus strain=OG-1\n",
      "GCF_002214465.1 Thermococcus barossii strain=SHCK-94\n",
      "GCF_014962245.1 Thermosphaera aggregans strain=3507\n",
      "GCF_002214605.1 Thermococcus chitonophagus strain=GC74\n",
      "Traceback (most recent call last):\n",
      "  File \"/tmp/ipykernel_83358/2651093835.py\", line 19, in <cell line: 8>\n",
      "    gfna_f.write(zlib.decompress(urllib.request.urlopen(ftp_prefix+\"_genomic.fna.gz\").read(), 15+32)) #\n",
      "  File \"/home/olga/miniconda3/envs/eagle/lib/python3.8/urllib/request.py\", line 222, in urlopen\n",
      "    return opener.open(url, data, timeout)\n",
      "  File \"/home/olga/miniconda3/envs/eagle/lib/python3.8/urllib/request.py\", line 525, in open\n",
      "    response = self._open(req, data)\n",
      "  File \"/home/olga/miniconda3/envs/eagle/lib/python3.8/urllib/request.py\", line 542, in _open\n",
      "    result = self._call_chain(self.handle_open, protocol, protocol +\n",
      "  File \"/home/olga/miniconda3/envs/eagle/lib/python3.8/urllib/request.py\", line 502, in _call_chain\n",
      "    result = func(*args)\n",
      "  File \"/home/olga/miniconda3/envs/eagle/lib/python3.8/urllib/request.py\", line 1397, in https_open\n",
      "    return self.do_open(http.client.HTTPSConnection, req,\n",
      "  File \"/home/olga/miniconda3/envs/eagle/lib/python3.8/urllib/request.py\", line 1354, in do_open\n",
      "    h.request(req.get_method(), req.selector, req.data, headers,\n",
      "  File \"/home/olga/miniconda3/envs/eagle/lib/python3.8/http/client.py\", line 1256, in request\n",
      "    self._send_request(method, url, body, headers, encode_chunked)\n",
      "  File \"/home/olga/miniconda3/envs/eagle/lib/python3.8/http/client.py\", line 1267, in _send_request\n",
      "    self.putrequest(method, url, **skips)\n",
      "  File \"/home/olga/miniconda3/envs/eagle/lib/python3.8/http/client.py\", line 1101, in putrequest\n",
      "    self._validate_path(url)\n",
      "  File \"/home/olga/miniconda3/envs/eagle/lib/python3.8/http/client.py\", line 1201, in _validate_path\n",
      "    raise InvalidURL(f\"URL can't contain control characters. {url!r} \"\n",
      "http.client.InvalidURL: URL can't contain control characters. '/genomes/all/GCF/900/012/635/GCF_900012635.1_Pyrococcus_chitonophagus_genome_sequence/GCF_900012635.1_Pyrococcus chitonophagus genome sequence_genomic.fna.gz' (found at least ' ')\n",
      "\n",
      "GCF_900012635.1 Thermococcus chitonophagus\n",
      "GCF_001433455.1 Thermococcus barophilus strain=CH5\n",
      "GCF_004803735.1 Halomicrobium mukohataei strain=JP60\n",
      "GCF_002813655.1 Methanobacterium subterraneum strain=MO-MB1\n",
      "GCF_002813695.1 Methanobacterium subterraneum strain=A8p\n",
      "GCF_010692885.1 Halogeometricum borinquense strain=wsp4\n",
      "GCF_004799645.1 Natronorubrum bangense strain=JCM 10635\n",
      "GCF_000006805.1 Halobacterium salinarum NRC-1 strain=NRC-1; ATCC 700922\n",
      "GCF_000009965.1 Thermococcus kodakarensis KOD1 strain=KOD1\n",
      "GCF_005890195.1 Natrinema pallidum strain=BOL6-1\n",
      "GCF_009729055.1 Sulfurisphaera ohwakuensis strain=TA-1\n",
      "GCF_000011105.1 Pyrococcus horikoshii OT3 strain=OT3\n",
      "GCF_002214385.1 Thermococcus gorgonarius strain=W-12\n",
      "GCF_002214485.1 Thermococcus pacificus strain=P-4\n",
      "GCF_002214505.1 Thermococcus siculi strain=RG-20\n",
      "GCF_002078355.1 Ferroplasma acidiphilum strain=Y\n",
      "GCF_000145295.1 Methanothermobacter marburgensis str. Marburg strain=Marburg\n",
      "Traceback (most recent call last):\n",
      "  File \"/tmp/ipykernel_83358/2651093835.py\", line 19, in <cell line: 8>\n",
      "    gfna_f.write(zlib.decompress(urllib.request.urlopen(ftp_prefix+\"_genomic.fna.gz\").read(), 15+32)) #\n",
      "  File \"/home/olga/miniconda3/envs/eagle/lib/python3.8/urllib/request.py\", line 222, in urlopen\n",
      "    return opener.open(url, data, timeout)\n",
      "  File \"/home/olga/miniconda3/envs/eagle/lib/python3.8/urllib/request.py\", line 525, in open\n",
      "    response = self._open(req, data)\n",
      "  File \"/home/olga/miniconda3/envs/eagle/lib/python3.8/urllib/request.py\", line 542, in _open\n",
      "    result = self._call_chain(self.handle_open, protocol, protocol +\n",
      "  File \"/home/olga/miniconda3/envs/eagle/lib/python3.8/urllib/request.py\", line 502, in _call_chain\n",
      "    result = func(*args)\n",
      "  File \"/home/olga/miniconda3/envs/eagle/lib/python3.8/urllib/request.py\", line 1397, in https_open\n",
      "    return self.do_open(http.client.HTTPSConnection, req,\n",
      "  File \"/home/olga/miniconda3/envs/eagle/lib/python3.8/urllib/request.py\", line 1354, in do_open\n",
      "    h.request(req.get_method(), req.selector, req.data, headers,\n",
      "  File \"/home/olga/miniconda3/envs/eagle/lib/python3.8/http/client.py\", line 1256, in request\n",
      "    self._send_request(method, url, body, headers, encode_chunked)\n",
      "  File \"/home/olga/miniconda3/envs/eagle/lib/python3.8/http/client.py\", line 1267, in _send_request\n",
      "    self.putrequest(method, url, **skips)\n",
      "  File \"/home/olga/miniconda3/envs/eagle/lib/python3.8/http/client.py\", line 1101, in putrequest\n",
      "    self._validate_path(url)\n",
      "  File \"/home/olga/miniconda3/envs/eagle/lib/python3.8/http/client.py\", line 1201, in _validate_path\n",
      "    raise InvalidURL(f\"URL can't contain control characters. {url!r} \"\n",
      "http.client.InvalidURL: URL can't contain control characters. '/genomes/all/GCF/900/036/045/GCF_900036045.1_Methanoculleus_sp_MAB1/GCF_900036045.1_Methanoculleus sp MAB1_genomic.fna.gz' (found at least ' ')\n",
      "\n",
      "GCF_900036045.1 Methanoculleus bourgensis\n",
      "GCF_005576615.1 Natrinema versiforme strain=BOL5-4\n",
      "GCF_001006045.1 Geoglobus ahangari strain=234\n",
      "GCF_010692905.1 Haloferax alexandrinus strain=wsp1\n",
      "GCF_900095295.1 Methanobacterium congolense\n",
      "GCF_019669965.1 Methanoculleus chikugoensis strain=MG62\n",
      "GCF_002214525.1 Thermococcus sp. P6 strain=P6\n",
      "GCF_000265525.1 Thermococcus cleftensis strain=CL1\n",
      "GCF_000007225.1 Pyrobaculum aerophilum str. IM2 strain=IM2\n",
      "GCF_022870485.1 Halococcus dombrowskii strain=H4\n",
      "GCF_000007305.1 Pyrococcus furiosus DSM 3638 strain=DSM 3638\n",
      "GCF_008245085.1 Pyrococcus furiosus DSM 3638 strain=DSM 3638\n",
      "GCF_000008645.1 Methanothermobacter thermautotrophicus str. Delta H strain=Delta H\n",
      "GCF_002214565.1 Thermococcus radiotolerans strain=EJ2\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "GCF_000007345.1 Methanosarcina acetivorans C2A strain=C2A\n",
      "GCF_000007185.1 Methanopyrus kandleri AV19 strain=AV19\n",
      "GCF_000007065.1 Methanosarcina mazei Go1 strain=Go1\n",
      "GCF_000585495.1 Thermococcus nautili strain=30-1\n",
      "GCF_000970205.1 Methanosarcina mazei S-6 strain=S-6\n",
      "GCF_021233435.1 Halobacterium noricense strain=JCM 15102\n",
      "GCF_000008665.1 Archaeoglobus fulgidus DSM 4304 strain=DSM 4304\n",
      "GCF_000404165.1 Methanobrevibacter sp. AbM4 strain=AbM4\n",
      "GCF_001477655.1 Methanobrevibacter millerae strain=SM9\n",
      "GCF_000091665.1 Methanocaldococcus jannaschii DSM 2661 strain=DSM 2661\n",
      "GCF_000151205.2 Thermococcus sp. AM4 strain=AM4\n",
      "GCF_000013725.1 Methanococcoides burtonii DSM 6242 strain=DSM 6242\n",
      "GCF_000008265.1 Picrophilus torridus DSM 9790 strain=DSM 9790\n",
      "GCF_000011585.1 Methanococcus maripaludis S2 strain=S2\n",
      "GCF_000591055.1 Natronomonas moolapensis 8.8.11 strain=8.8.11\n",
      "GCF_000011125.1 Aeropyrum pernix K1 strain=K1\n",
      "GCF_000011085.1 Haloarcula marismortui ATCC 43049 strain=ATCC 43049\n",
      "GCF_005310945.1 Haloarcula marismortui ATCC 43049 strain=ATCC 43049\n",
      "GCF_000195935.2 Pyrococcus abyssi GE5 strain=Orsay\n",
      "GCF_000007005.1 Saccharolobus solfataricus P2 strain=P2\n",
      "GCF_000011205.1 Sulfurisphaera tokodaii str. 7 strain=7\n",
      "GCF_000195915.1 Thermoplasma acidophilum DSM 1728 strain=DSM 1728\n",
      "GCF_000011185.1 Thermoplasma volcanium GSS1 strain=GSS1\n",
      "GCF_002214545.1 Thermococcus thioreducens strain=OGL-20P\n",
      "GCF_002116695.1 Acidianus manzaensis strain=YN-25\n",
      "GCF_001563245.1 Methanobrevibacter olleyae strain=YLM1\n",
      "GCF_000011005.1 Methanocella paludicola SANAE strain=SANAE\n",
      "GCF_000025685.1 Haloferax volcanii DS2 strain=DS2\n",
      "GCF_000013445.1 Methanospirillum hungatei JF-1 strain=JF-1\n",
      "GCF_000012285.1 Sulfolobus acidocaldarius DSM 639 strain=DSM 639\n",
      "GCF_000152265.2 Ferroplasma acidarmanus fer1 strain=fer1\n",
      "GCF_004126515.1 Halorubrum ezzemoulense strain=Fb21\n",
      "GCF_000012545.1 Methanosphaera stadtmanae DSM 3091 strain=DSM 3091\n",
      "GCF_000016385.1 Pyrobaculum arsenaticum DSM 13514 strain=DSM 13514\n",
      "GCF_000211475.1 Pyrococcus sp. NA2 strain=NA2\n",
      "GCF_000026045.1 Natronomonas pharaonis DSM 2160 strain=Gabara\n",
      "GCF_009392895.1 Natronorubrum aibiense strain=7-3\n",
      "GCF_000014945.1 Methanothrix thermoacetophila PT strain=PT\n",
      "GCF_000063445.1 Methanocella arvoryzae MRE50 strain=MRE50\n",
      "GCF_000226975.2 Halobiforma lacisalsi AJ5 strain=AJ5\n",
      "GCF_000009185.1 Haloquadratum walsbyi DSM 16790 strain=DSM 16790 = HBSQ001\n",
      "GCF_000015825.1 Methanoculleus marisnigri JR1 strain=JR1\n",
      "GCF_000015225.1 Thermofilum pendens Hrk 5 strain=Hrk 5\n",
      "GCF_017352095.1 Haloterrigena longa strain=ABH32\n",
      "GCF_000019605.1 Candidatus Korarchaeum cryptofilum OPF8\n",
      "GCF_000015205.1 Pyrobaculum islandicum DSM 4184 strain=DSM 4184\n",
      "GCF_000385565.1 Archaeoglobus sulfaticallidus PM70-1 strain=PM70-1\n",
      "GCF_000151105.2 Thermococcus barophilus MP strain=MP\n",
      "GCF_000018305.1 Caldivirga maquilingensis IC-167 strain=IC-167\n",
      "GCF_000015945.1 Staphylothermus marinus F1 strain=F1\n",
      "GCF_000016125.1 Methanococcus maripaludis C5 strain=C5\n",
      "GCF_000017165.1 Methanococcus vannielii SB strain=SB\n",
      "GCF_000281695.1 Natrinema sp. J7-2 strain=J7-2\n",
      "GCF_000015765.1 Methanocorpusculum labreanum Z strain=Z\n",
      "GCF_000015805.1 Pyrobaculum calidifontis JCM 11548 strain=JCM 11548\n",
      "GCF_000015145.1 Hyperthermus butylicus DSM 5456 strain=DSM 5456\n",
      "GCF_000022205.1 Halorubrum lacusprofundi ATCC 49239 strain=ATCC 49239\n",
      "GCF_000017185.1 Methanococcus aeolicus Nankai-3 strain=Nankai-3\n",
      "GCF_000022485.1 Sulfolobus islandicus Y.N.15.51 strain=Y.N.15.51\n",
      "GCF_000016525.1 Methanobrevibacter smithii ATCC 35061 strain=ATCC 35061; PS; DSMZ 861\n",
      "GCF_000022445.1 Sulfolobus islandicus M.16.4 strain=M.16.4\n",
      "GCF_000022405.1 Sulfolobus islandicus M.14.25 strain=M.14.25\n",
      "GCF_000022425.1 Sulfolobus islandicus M.16.27 strain=M.16.27\n",
      "GCF_000022385.1 Sulfolobus islandicus L.S.2.15 strain=L.S.2.15\n",
      "GCF_000018465.1 Nitrosopumilus maritimus SCM1 strain=SCM1\n",
      "GCF_000022465.1 Sulfolobus islandicus Y.G.57.14 strain=Y.G.57.14\n",
      "GCF_000025665.1 Aciduliprofundum boonei T469 strain=T469\n",
      "GCF_000019805.1 Pyrobaculum neutrophilum V24Sta strain=V24Sta\n",
      "GCF_000017945.1 Ignicoccus hospitalis KIN4/I strain=KIN4/I\n",
      "GCF_000017625.1 Methanoregula boonei 6A8 strain=6A8\n",
      "GCF_000172995.2 Halogeometricum borinquense DSM 11551 strain=DSM 11551\n",
      "GCF_000069025.1 Halobacterium salinarum R1 strain=DSM 671 = R1\n",
      "GCF_000023965.1 Halomicrobium mukohataei DSM 12286 strain=DSM 12286\n",
      "GCF_000020905.1 Desulfurococcus amylolyticus 1221n strain=1221n\n",
      "GCF_000023945.1 Halorhabdus utahensis DSM 12940 strain=DSM 12940\n",
      "GCF_000021965.1 Methanosphaerula palustris E1-9c strain=E1-9c\n",
      "GCF_000306765.2 Haloferax mediterranei ATCC 33500 strain=CGMCC 1.2087\n",
      "GCF_005406325.1 Haloferax mediterranei ATCC 33500 strain=ATCC 33500\n",
      "GCF_000969885.1 Methanosarcina thermophila TM-1 strain=TM-1\n",
      "GCF_000166095.1 Methanothermus fervidus DSM 2088 strain=DSM 2088\n",
      "GCF_019175345.1 Saccharolobus shibatae B12 strain=B12\n",
      "GCF_000246985.2 Thermococcus litoralis DSM 5473 strain=DSM 5473\n",
      "GCF_000018365.1 Thermococcus onnurineus NA1 strain=NA1\n",
      "GCF_000215995.1 Pyrococcus yayanosii CH1 strain=CH1\n",
      "GCF_013388255.1 Methanolobus zinderi strain=DSM 21339\n",
      "GCF_000025325.1 Haloterrigena turkmenica DSM 5511 strain=DSM 5511\n",
      "GCF_000025865.1 Methanohalophilus mahii DSM 5219 strain=DSM 5219\n",
      "GCF_000025625.1 Natrialba magadii ATCC 43099 strain=ATCC 43099\n",
      "GCF_020614395.1 Salarchaeum japonicum strain=JCM 16327\n",
      "GCF_000148385.1 Vulcanisaeta distributa DSM 14429 strain=DSM 14429\n",
      "GCF_000025285.1 Archaeoglobus profundus DSM 5631 strain=DSM 5631\n",
      "GCF_000092305.1 Methanocaldococcus infernus ME strain=ME\n",
      "GCF_000023985.1 Methanocaldococcus fervens AG86 strain=AG86\n",
      "GCF_000024625.1 Methanocaldococcus vulcanius M7 strain=M7\n",
      "GCF_000517445.1 Thermococcus paralvinellae strain=ES1\n",
      "GCF_001971705.1 Haloterrigena daqingensis strain=JX313\n",
      "GCF_000025505.1 Ferroglobus placidus DSM 10642 strain=DSM 10642\n",
      "GCF_000092465.1 Staphylothermus hellenicus DSM 12710 strain=DSM 12710\n",
      "GCF_000022365.1 Thermococcus gammatolerans EJ3 strain=EJ3; DSM 15229\n",
      "GCF_000327485.1 Methanoregula formicica SMSP strain=SMSP\n",
      "GCF_000022545.1 Thermococcus sibiricus MM 739 strain=MM 739\n",
      "GCF_003201765.2 Acidianus sulfidivorans JP7 strain=JP7\n",
      "GCF_000092185.1 Thermosphaera aggregans DSM 11486 strain=DSM 11486\n",
      "GCF_003990725.2 Halorubrum sp. PV6 strain=PV6\n",
      "GCF_000223905.1 Haloarcula hispanica ATCC 33960 strain=CGMCC 1.2049\n",
      "GCF_000024185.1 Methanobrevibacter ruminantium M1 strain=M1\n",
      "GCF_000828535.1 Methanococcus maripaludis KA1 strain=KA1\n",
      "GCF_000828555.1 Methanococcus maripaludis OS7 strain=OS7\n",
      "GCF_000025525.1 Methanocaldococcus sp. FS406-22 strain=FS406-22\n",
      "GCF_000196655.1 Methanohalobium evestigatum Z-7303 strain=Z-7303\n",
      "GCF_000179575.2 Methanothermococcus okinawensis IH1 strain=IH1\n",
      "GCF_003856835.1 Haloplanus aerogenes strain=JCM 16430\n",
      "GCF_018200015.1 Haloarcula sinaiiensis ATCC 33800 strain=ATCC 33800\n",
      "GCF_000144915.1 Acidilobus saccharovorans 345-15 strain=345-15\n",
      "GCF_000217995.1 Methanosalsum zhilinae DSM 4017 strain=DSM 4017\n",
      "GCF_000147875.1 Methanolacinia petrolearia DSM 11571 strain=DSM 11571\n",
      "GCF_023167465.1 Methanothermobacter tenebrarum strain=RMAS\n",
      "GCF_000194625.1 Archaeoglobus veneficus SNP6 strain=SNP6\n",
      "GCF_000223395.1 Pyrolobus fumarii 1A strain=1A\n",
      "GCF_000328685.1 Natronococcus occultus SP4 strain=SP4\n",
      "GCF_000813245.1 Thermofilum adornatus 1505 strain=1505\n",
      "GCF_004799685.1 Halobellus limi strain=CGMCC 1.10331\n",
      "GCF_002156705.1 Natrarchaeobaculum aegyptiacum strain=JW/NM-HA 15\n",
      "GCF_000230955.2 Halobacterium sp. DL1 strain=DL1\n",
      "GCF_002906575.1 Salinigranum rubrum strain=GX10\n",
      "GCF_000186365.1 Desulfurococcus mucosus DSM 2162 strain=DSM 2162\n",
      "GCF_000237865.1 Haloquadratum walsbyi C23 strain=DSM 16854 = C23\n",
      "GCF_000231015.2 Desulfurococcus amylolyticus DSM 16532 strain=DSM 16532\n",
      "GCF_000253055.1 Thermoproteus tenax Kra 1 strain=Kra 1\n",
      "GCF_000196895.1 Halalkalicoccus jeotgali B3 strain=B3\n",
      "GCF_001027005.1 Methanosarcina barkeri CM1 strain=CM1\n",
      "GCF_000217715.1 Halopiger xanaduensis SH-6 strain=SH-6\n",
      "GCF_000517625.1 Halostagnicola larsenii XH-48 strain=XH-48\n",
      "GCF_000328525.1 Halovivax ruber XH-70 strain=XH-70\n",
      "GCF_000230735.2 Natrinema pellirubrum DSM 15624 strain=DSM 15624\n",
      "GCF_000230715.2 Natronobacterium gregoryi SP2 strain=SP2\n",
      "GCF_000828575.1 Methanothermobacter sp. CaT2 strain=CaT2\n",
      "GCF_000328665.1 Methanomethylovorans hollandica DSM 15978 strain=DSM 15978\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "GCF_000214725.1 Methanobacterium paludis strain=SWAN1\n",
      "GCF_013415905.1 Halosimplex pelagicum strain=R2\n",
      "GCF_013415885.1 Halosimplex rubrum strain=R27\n",
      "GCF_000191585.1 Methanobacterium lacus strain=AL-21\n",
      "GCF_000214415.1 Methanotorris igneus Kol 5 strain=Kol 5\n",
      "GCF_000698785.1 Nitrososphaera viennensis EN76 strain=EN76\n",
      "GCF_000189575.1 Sulfolobus islandicus HVE10/4 strain=HVE10/4\n",
      "GCF_000189555.1 Sulfolobus islandicus REY15A strain=REY15A\n",
      "GCF_000213215.1 Acidianus hospitalis W1 strain=W1\n",
      "GCF_001481685.1 Ignicoccus islandicus DSM 13165 strain=DSM 13165\n",
      "GCF_000190315.1 Vulcanisaeta moutnovskia 768-28 strain=768-28\n",
      "GCF_000204415.1 Methanothrix soehngenii GP6 strain=GP6\n",
      "GCF_000193375.1 Thermoproteus uzoniensis 768-20 strain=768-20\n",
      "GCF_000204925.1 Metallosphaera cuprina Ar-4 strain=Ar-4\n",
      "GCF_020700235.1 Haladaptatus pallidirubidus strain=YIM 93656\n",
      "GCF_019823105.1 Halobaculum magnesiiphilum strain=NBRC 109044\n",
      "GCF_000340315.1 Sulfolobus acidocaldarius N8 strain=N8\n",
      "GCF_000338775.1 Sulfolobus acidocaldarius Ron12/I strain=Ron12/I\n",
      "GCF_000470655.1 Halorhabdus tiamatea SARL4B strain=type strain: SARL4B\n",
      "GCF_000251105.1 Methanocella conradii HZ254 strain=HZ254\n",
      "GCF_000221185.1 Thermococcus sp. 4557 strain=4557\n",
      "GCF_002355635.1 Halopenitus persicus strain=CBA1233\n",
      "GCF_000220645.1 Methanococcus maripaludis X1 strain=X1\n",
      "GCF_000317795.1 Caldisphaera lagunensis DSM 15908 strain=DSM 15908\n",
      "GCF_000234805.1 Pyrobaculum ferrireducens strain=1860\n",
      "GCF_000235565.1 Methanosaeta harundinacea 6Ac strain=6Ac\n",
      "GCF_000258425.1 Fervidicoccus fontis Kam940 strain=Kam940\n",
      "GCF_000263735.1 Pyrococcus sp. ST04 strain=ST04\n",
      "GCF_000264495.1 Thermogladius calderae 1633 strain=1633\n",
      "GCF_000275605.1 Pyrococcus furiosus COM1 strain=COM1\n",
      "GCF_016065055.1 Halosimplex litoreum strain=YGH94\n",
      "GCF_000591035.1 Aeropyrum camini SY1 = JCM 12091 strain=SY1\n",
      "GCF_000304355.2 Methanoculleus bourgensis MS2 strain=type strain:MS2\n",
      "GCF_000299365.1 Candidatus Nitrosopumilus koreensis AR1 strain=AR1\n",
      "GCF_000299395.1 Candidatus Nitrosopumilus sediminis strain=AR2\n",
      "GCF_000300255.2 Candidatus Methanomethylophilus alvus Mx1201 strain=Mx1201\n",
      "GCF_000341715.1 Methanosarcina mazei Tuc01 strain=Tuc01\n",
      "GCF_000303155.1 Candidatus Nitrososphaera gargensis Ga9.2\n",
      "GCF_000364745.1 Sulfolobus islandicus LAL14/1 strain=LAL14/1\n",
      "GCF_001412615.1 Pyrodictium delaneyi strain=Su06\n",
      "GCF_003711245.1 Candidatus Methanomethylophilus alvus strain=Mx-05\n",
      "GCF_902387285.1 Candidatus Methanomethylophilus alvus\n",
      "GCF_003201675.2 Metallosphaera hakonensis JCM 8857 = DSM 7519 strain=HO1-1\n",
      "GCF_002214365.1 Thermococcus celer Vu 13 = JCM 8558 strain=Vu 13\n",
      "GCF_008326385.1 Sulfuracidifex tepidarius strain=IC-007\n",
      "GCF_008326425.1 Sulfuracidifex tepidarius strain=IC-006\n",
      "GCF_000404225.1 Candidatus Methanomassiliicoccus intestinalis Issoire-Mx1 strain=Issoire-Mx1\n",
      "GCF_000739065.1 Methanocaldococcus bathoardescens strain=JH146\n",
      "GCF_000403645.1 Salinarchaeum sp. Harcht-Bsk1 strain=Harcht-Bsk1\n",
      "GCF_000725425.1 Palaeococcus pacificus DY20341 strain=DY20341\n",
      "GCF_000734035.1 Archaeoglobus fulgidus DSM 8774 strain=DSM 8774\n",
      "GCF_000802205.1 Candidatus Nitrosocosmicus oleophilus strain=MY3\n",
      "GCF_000446015.1 Thermofilum adornatum strain=1910b\n",
      "GCF_009690625.1 Halorhabdus sp. CBA1104 strain=CBA1104\n",
      "GCF_902383905.1 Candidatus Methanomassiliicoccus intestinalis\n",
      "Traceback (most recent call last):\n",
      "  File \"/tmp/ipykernel_83358/2651093835.py\", line 19, in <cell line: 8>\n",
      "    gfna_f.write(zlib.decompress(urllib.request.urlopen(ftp_prefix+\"_genomic.fna.gz\").read(), 15+32)) #\n",
      "  File \"/home/olga/miniconda3/envs/eagle/lib/python3.8/urllib/request.py\", line 222, in urlopen\n",
      "    return opener.open(url, data, timeout)\n",
      "  File \"/home/olga/miniconda3/envs/eagle/lib/python3.8/urllib/request.py\", line 525, in open\n",
      "    response = self._open(req, data)\n",
      "  File \"/home/olga/miniconda3/envs/eagle/lib/python3.8/urllib/request.py\", line 542, in _open\n",
      "    result = self._call_chain(self.handle_open, protocol, protocol +\n",
      "  File \"/home/olga/miniconda3/envs/eagle/lib/python3.8/urllib/request.py\", line 502, in _call_chain\n",
      "    result = func(*args)\n",
      "  File \"/home/olga/miniconda3/envs/eagle/lib/python3.8/urllib/request.py\", line 1397, in https_open\n",
      "    return self.do_open(http.client.HTTPSConnection, req,\n",
      "  File \"/home/olga/miniconda3/envs/eagle/lib/python3.8/urllib/request.py\", line 1354, in do_open\n",
      "    h.request(req.get_method(), req.selector, req.data, headers,\n",
      "  File \"/home/olga/miniconda3/envs/eagle/lib/python3.8/http/client.py\", line 1256, in request\n",
      "    self._send_request(method, url, body, headers, encode_chunked)\n",
      "  File \"/home/olga/miniconda3/envs/eagle/lib/python3.8/http/client.py\", line 1267, in _send_request\n",
      "    self.putrequest(method, url, **skips)\n",
      "  File \"/home/olga/miniconda3/envs/eagle/lib/python3.8/http/client.py\", line 1101, in putrequest\n",
      "    self._validate_path(url)\n",
      "  File \"/home/olga/miniconda3/envs/eagle/lib/python3.8/http/client.py\", line 1201, in _validate_path\n",
      "    raise InvalidURL(f\"URL can't contain control characters. {url!r} \"\n",
      "http.client.InvalidURL: URL can't contain control characters. '/genomes/all/GCF/001/488/575/GCF_001488575.1_Halobacterium_hubeiense_JI20-1/GCF_001488575.1_Halobacterium hubeiense JI20-1_genomic.fna.gz' (found at least ' ')\n",
      "\n",
      "GCF_001488575.1 Halobacterium hubeiense strain=JI20-1\n",
      "GCF_000812185.1 Candidatus Nitrosopelagicus brevis strain=CN25\n",
      "GCF_000504565.1 Haloarcula hispanica N601 strain=N601\n",
      "GCF_023008585.1 Natribaculum breve strain=TRM20010\n",
      "GCF_023008565.1 Natribaculum longum strain=TRM20345\n",
      "GCF_000816105.1 Thermococcus guaymasensis DSM 11113 strain=DSM 11113\n",
      "GCF_000969945.1 Methanosarcina sp. Kolksee strain=Kolksee\n",
      "GCF_000970045.1 Methanosarcina sp. MTP4 strain=MTP4\n",
      "GCF_000970005.1 Methanosarcina sp. WH1 strain=WH1\n",
      "GCF_000969965.1 Methanosarcina sp. WWM596 strain=WWM596\n",
      "GCF_000970325.1 Methanococcoides methylutens MM1 strain=MM1\n",
      "GCF_000970065.1 Methanosarcina barkeri 227 strain=227\n",
      "GCF_000970305.1 Methanosarcina barkeri 3 strain=3\n",
      "GCF_000970025.1 Methanosarcina barkeri MS strain=MS\n",
      "GCF_000969985.1 Methanosarcina barkeri str. Wiesmoor strain=Wiesmoor\n",
      "GCF_000970285.1 Methanosarcina horonobensis HB-1 = JCM 15518 strain=HB-1\n",
      "GCF_000970265.1 Methanosarcina lacustris Z-7289 strain=Z-7289\n",
      "GCF_000970245.1 Methanosarcina mazei C16 strain=C16\n",
      "GCF_000970225.1 Methanosarcina mazei LYC strain=LYC\n",
      "GCF_000970185.1 Methanosarcina mazei SarPi strain=SarPi\n",
      "GCF_000970165.1 Methanosarcina mazei WWM610 strain=WWM610\n",
      "GCF_000970145.1 Methanosarcina siciliae C2J strain=C2J\n",
      "GCF_000970125.1 Methanosarcina siciliae HI350 strain=HI350\n",
      "GCF_000970085.1 Methanosarcina siciliae T4/M strain=T4/M\n",
      "GCF_000969925.1 Methanosarcina thermophila CHTI-55 strain=CHTI-55\n",
      "GCF_000969905.1 Methanosarcina vacuolata Z-761 strain=Z-761\n",
      "GCF_023093535.1 Halovivax limisalsi strain=IBRC-M 10022\n",
      "GCF_004799665.1 Halapricum salinum strain=CBA1105\n",
      "GCF_000730285.1 Candidatus Nitrososphaera evergladensis SR1\n",
      "GCF_013407145.1 Nitrosopumilus cobalaminigenes strain=HCA1\n",
      "GCF_013407185.1 Nitrosopumilus ureiphilus strain=PS0\n",
      "GCF_000769655.1 Thermococcus eurythermalis strain=A501\n",
      "GCF_013343295.1 Metallosphaera tengchongensis strain=Ric-A\n",
      "GCF_003342675.1 Haloplanus rubicundus strain=CBA1112\n",
      "GCF_003342695.1 Haloplanus rubicundus strain=CBA1113\n",
      "GCF_000993805.1 Infirmifilum uzonense strain=1807-2\n",
      "GCF_003431325.1 Acidilobus sp. 7A strain=7A\n",
      "GCF_000800805.1 Candidatus Methanoplasma termitum strain=MpT1\n",
      "GCF_000956175.1 Nitrosopumilus adriaticus strain=NF5\n",
      "GCF_000875775.1 Nitrosopumilus piranensis strain=D3C\n",
      "GCF_023008545.1 Halovarius luteus strain=IBRC-M 10912\n",
      "GCF_020217425.1 Halosiccatus urmianus strain=IBRC-M: 10911\n",
      "GCF_000827835.1 Haloarcula sp. CBA1115 strain=CBA1115\n",
      "GCF_000955905.1 Candidatus Nitrosotenuis cloacae strain=SAT1\n",
      "GCF_001011115.1 Halanaeroarchaeum sulfurireducens strain=HSR2\n",
      "GCF_001305655.1 Halanaeroarchaeum sulfurireducens strain=M27-SA2\n",
      "GCF_001577775.1 Pyrococcus kukulkanii strain=NCB100\n",
      "GCF_001548675.1 Methanobrevibacter sp. YE315 strain=YE315\n",
      "GCF_003967175.1 Sulfodiicoccus acidiphilus strain=HS-1\n",
      "GCF_900083515.1 Cuniculiplasma divulgatum strain=S5(T) (=JCM 30642; =VKM B-2941)\n",
      "Traceback (most recent call last):\n",
      "  File \"/tmp/ipykernel_83358/2651093835.py\", line 19, in <cell line: 8>\n",
      "    gfna_f.write(zlib.decompress(urllib.request.urlopen(ftp_prefix+\"_genomic.fna.gz\").read(), 15+32)) #\n",
      "  File \"/home/olga/miniconda3/envs/eagle/lib/python3.8/urllib/request.py\", line 222, in urlopen\n",
      "    return opener.open(url, data, timeout)\n",
      "  File \"/home/olga/miniconda3/envs/eagle/lib/python3.8/urllib/request.py\", line 525, in open\n",
      "    response = self._open(req, data)\n",
      "  File \"/home/olga/miniconda3/envs/eagle/lib/python3.8/urllib/request.py\", line 542, in _open\n",
      "    result = self._call_chain(self.handle_open, protocol, protocol +\n",
      "  File \"/home/olga/miniconda3/envs/eagle/lib/python3.8/urllib/request.py\", line 502, in _call_chain\n",
      "    result = func(*args)\n",
      "  File \"/home/olga/miniconda3/envs/eagle/lib/python3.8/urllib/request.py\", line 1397, in https_open\n",
      "    return self.do_open(http.client.HTTPSConnection, req,\n",
      "  File \"/home/olga/miniconda3/envs/eagle/lib/python3.8/urllib/request.py\", line 1354, in do_open\n",
      "    h.request(req.get_method(), req.selector, req.data, headers,\n",
      "  File \"/home/olga/miniconda3/envs/eagle/lib/python3.8/http/client.py\", line 1256, in request\n",
      "    self._send_request(method, url, body, headers, encode_chunked)\n",
      "  File \"/home/olga/miniconda3/envs/eagle/lib/python3.8/http/client.py\", line 1267, in _send_request\n",
      "    self.putrequest(method, url, **skips)\n",
      "  File \"/home/olga/miniconda3/envs/eagle/lib/python3.8/http/client.py\", line 1101, in putrequest\n",
      "    self._validate_path(url)\n",
      "  File \"/home/olga/miniconda3/envs/eagle/lib/python3.8/http/client.py\", line 1201, in _validate_path\n",
      "    raise InvalidURL(f\"URL can't contain control characters. {url!r} \"\n",
      "http.client.InvalidURL: URL can't contain control characters. '/genomes/all/GCF/900/090/055/GCF_900090055.1_C.divulgatum_PM4/GCF_900090055.1_C.divulgatum PM4_genomic.fna.gz' (found at least ' ')\n",
      "\n",
      "GCF_900090055.1 Cuniculiplasma divulgatum strain=PM4 (=JCM 30641; =VKM B-2940)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "GCF_001484685.1 Thermococcus sp. 2319x1 strain=2319x1\n",
      "GCF_902813195.1 Thermococcus sp. 2319x1 strain=Essen\n",
      "GCF_003058365.1 Halococcoides cellulosivorans strain=HArcel1T\n",
      "GCF_013402815.2 Natrinema halophilum strain=YPL8\n",
      "GCF_020886315.1 Saccharolobus caldissimus strain=JCM 32116\n",
      "GCF_013456555.2 Natrinema sp. YPL30 strain=YPL30\n",
      "GCF_013391105.1 Natronomonas salina strain=YPL13\n",
      "GCF_001647085.1 Thermococcus piezophilus strain=CDGS\n",
      "GCF_001304615.2 Methanosarcina flavescens strain=E03.2\n",
      "GCF_003268005.1 Methanosphaera sp. BMS strain=BMS\n",
      "Traceback (most recent call last):\n",
      "  File \"/tmp/ipykernel_83358/2651093835.py\", line 19, in <cell line: 8>\n",
      "    gfna_f.write(zlib.decompress(urllib.request.urlopen(ftp_prefix+\"_genomic.fna.gz\").read(), 15+32)) #\n",
      "  File \"/home/olga/miniconda3/envs/eagle/lib/python3.8/urllib/request.py\", line 222, in urlopen\n",
      "    return opener.open(url, data, timeout)\n",
      "  File \"/home/olga/miniconda3/envs/eagle/lib/python3.8/urllib/request.py\", line 525, in open\n",
      "    response = self._open(req, data)\n",
      "  File \"/home/olga/miniconda3/envs/eagle/lib/python3.8/urllib/request.py\", line 542, in _open\n",
      "    result = self._call_chain(self.handle_open, protocol, protocol +\n",
      "  File \"/home/olga/miniconda3/envs/eagle/lib/python3.8/urllib/request.py\", line 502, in _call_chain\n",
      "    result = func(*args)\n",
      "  File \"/home/olga/miniconda3/envs/eagle/lib/python3.8/urllib/request.py\", line 1397, in https_open\n",
      "    return self.do_open(http.client.HTTPSConnection, req,\n",
      "  File \"/home/olga/miniconda3/envs/eagle/lib/python3.8/urllib/request.py\", line 1354, in do_open\n",
      "    h.request(req.get_method(), req.selector, req.data, headers,\n",
      "  File \"/home/olga/miniconda3/envs/eagle/lib/python3.8/http/client.py\", line 1256, in request\n",
      "    self._send_request(method, url, body, headers, encode_chunked)\n",
      "  File \"/home/olga/miniconda3/envs/eagle/lib/python3.8/http/client.py\", line 1267, in _send_request\n",
      "    self.putrequest(method, url, **skips)\n",
      "  File \"/home/olga/miniconda3/envs/eagle/lib/python3.8/http/client.py\", line 1101, in putrequest\n",
      "    self._validate_path(url)\n",
      "  File \"/home/olga/miniconda3/envs/eagle/lib/python3.8/http/client.py\", line 1201, in _validate_path\n",
      "    raise InvalidURL(f\"URL can't contain control characters. {url!r} \"\n",
      "http.client.InvalidURL: URL can't contain control characters. '/genomes/all/GCF/900/696/045/GCF_900696045.1_Nitrosocosmicus_franklandus_C13/GCF_900696045.1_Nitrosocosmicus franklandus C13_genomic.fna.gz' (found at least ' ')\n",
      "\n",
      "GCF_900696045.1 Candidatus Nitrosocosmicus franklandus\n",
      "GCF_009762275.1 Haloplanus rallus strain=MBLA0036\n",
      "GCF_001870125.1 Candidatus Nitrosocosmicus hydrocola strain=G61\n",
      "GCF_002787055.1 Candidatus Nitrosotenuis aquarius strain=AQ6F\n",
      "GCF_003721175.2 Halarchaeum sp. CBA1220 strain=CBA1220\n",
      "GCF_002025255.2 Halolamina sp. CBA1230 strain=CBA1230\n",
      "GCF_003721435.2 Halorubrum sp. CBA1229 strain=CBA1229\n",
      "GCF_001767315.1 Halodesulfurarchaeum formicicum strain=HTSR1\n",
      "GCF_001886955.1 Halodesulfurarchaeum formicicum strain=HSR6\n",
      "GCF_013340765.1 Conexivisphaera calida strain=NAS-02\n",
      "GCF_001719125.1 Sulfolobus sp. A20 strain=A20\n",
      "GCF_002156965.1 Candidatus Nitrosomarinus catalina strain=SPOT01\n",
      "GCF_002813675.1 Methanobacterium sp. MZ-A1 strain=MZ-A1\n",
      "GCF_002214165.1 Candidatus Mancarchaeum acidiphilum strain=Mia14\n",
      "GCF_002844335.1 Haloarcula taiwanensis strain=Taiwanensis\n",
      "GCF_001989615.1 Halorientalis sp. IM1011 strain=IM1011\n",
      "GCF_013407165.1 Nitrosopumilus oxyclinae strain=HCE1\n",
      "GCF_002197185.1 Thermococcus sp. 5-4 strain=5-4\n",
      "GCF_900198835.1 Thermococcus henrietii strain=EXT12c\n",
      "GCF_904067545.1 Thermococcus camini strain=IRI35c\n",
      "GCF_021233415.1 Halobacterium litoreum strain=ZS-54-S2\n",
      "GCF_003430805.1 Natrarchaeobaculum sulfurireducens strain=AArc-Mg\n",
      "GCF_003430825.1 Natrarchaeobaculum sulfurireducens strain=AArc1\n",
      "GCF_002906215.1 Candidatus Nitrosocaldus islandicus\n",
      "GCF_002952775.1 Halalkaliarchaeum desulfuricum strain=AArc-Sl\n",
      "Traceback (most recent call last):\n",
      "  File \"/tmp/ipykernel_83358/2651093835.py\", line 19, in <cell line: 8>\n",
      "    gfna_f.write(zlib.decompress(urllib.request.urlopen(ftp_prefix+\"_genomic.fna.gz\").read(), 15+32)) #\n",
      "  File \"/home/olga/miniconda3/envs/eagle/lib/python3.8/urllib/request.py\", line 222, in urlopen\n",
      "    return opener.open(url, data, timeout)\n",
      "  File \"/home/olga/miniconda3/envs/eagle/lib/python3.8/urllib/request.py\", line 525, in open\n",
      "    response = self._open(req, data)\n",
      "  File \"/home/olga/miniconda3/envs/eagle/lib/python3.8/urllib/request.py\", line 542, in _open\n",
      "    result = self._call_chain(self.handle_open, protocol, protocol +\n",
      "  File \"/home/olga/miniconda3/envs/eagle/lib/python3.8/urllib/request.py\", line 502, in _call_chain\n",
      "    result = func(*args)\n",
      "  File \"/home/olga/miniconda3/envs/eagle/lib/python3.8/urllib/request.py\", line 1397, in https_open\n",
      "    return self.do_open(http.client.HTTPSConnection, req,\n",
      "  File \"/home/olga/miniconda3/envs/eagle/lib/python3.8/urllib/request.py\", line 1354, in do_open\n",
      "    h.request(req.get_method(), req.selector, req.data, headers,\n",
      "  File \"/home/olga/miniconda3/envs/eagle/lib/python3.8/http/client.py\", line 1256, in request\n",
      "    self._send_request(method, url, body, headers, encode_chunked)\n",
      "  File \"/home/olga/miniconda3/envs/eagle/lib/python3.8/http/client.py\", line 1267, in _send_request\n",
      "    self.putrequest(method, url, **skips)\n",
      "  File \"/home/olga/miniconda3/envs/eagle/lib/python3.8/http/client.py\", line 1101, in putrequest\n",
      "    self._validate_path(url)\n",
      "  File \"/home/olga/miniconda3/envs/eagle/lib/python3.8/http/client.py\", line 1201, in _validate_path\n",
      "    raise InvalidURL(f\"URL can't contain control characters. {url!r} \"\n",
      "http.client.InvalidURL: URL can't contain control characters. '/genomes/all/GCF/900/248/165/GCF_900248165.1_Nitrosocaldus_cavascurensis_SCU2_chromosome/GCF_900248165.1_Nitrosocaldus cavascurensis SCU2 chromosome_genomic.fna.gz' (found at least ' ')\n",
      "\n",
      "GCF_900248165.1 Candidatus Nitrosocaldus cavascurensis strain=SCU2\n",
      "GCF_019880245.1 Halobaculum roseum strain=CGMCC 1.15501\n",
      "GCF_013407275.1 Candidatus Nitrosotenuis sp. DW1 strain=DW1\n",
      "GCF_013407385.1 Nitrosarchaeum sp. AC2 strain=AC2\n",
      "GCF_016806735.1 Candidatus Micrarchaeum sp. strain=A_DKE\n",
      "GCF_004114375.1 Halorubrum sp. BOL3-1 strain=BOL3-1\n",
      "GCF_004087835.1 Halorussus pelagicus strain=RC-68\n",
      "GCF_006740685.1 Candidatus Nitrosopumilus sp. SW strain=SW\n",
      "GCF_902827225.1 Methanocaldococcus sp. SG7 strain=SG7\n",
      "GCF_009617975.1 Candidatus Nanohalobium constans strain=LC1Nh\n",
      "GCF_005155585.1 Haloprofundus sp. MHR1 strain=MHR1\n",
      "GCF_006274605.1 Thermococcus indicus strain=IOH1\n",
      "GCF_023169565.1 Metallosphaera sp. MJ1HA strain=MJ1HA\n",
      "GCF_008000775.1 Candidatus Prometheoarchaeum syntrophicum strain=MK-D1\n",
      "GCF_008152015.1 Thermococcus aciditolerans strain=SY113\n",
      "GCF_007833275.1 Salarchaeum sp. JOR-1 strain=JOR-1\n",
      "GCF_008033705.1 Methanothermobacter sp. KEPCO-1 strain=KEPCO-1\n",
      "GCF_009914355.1 Methanothermobacter sp. THM-1 strain=THM-1\n",
      "GCF_009917665.1 Methanothermobacter sp. THM-2\n",
      "GCF_003351865.1 Methanofervidicoccus sp. A16 strain=A16\n",
      "GCF_009617995.1 Halomicrobium sp. LC1Hm strain=LC1Hm\n",
      "GCF_008831545.1 Halorussus halophilus strain=ZS-3\n",
      "GCF_009601705.1 Sulfolobus sp. E5-1-F strain=E5-1-F\n",
      "GCF_009602405.1 Sulfolobus sp. E11-6 strain=E11-6\n",
      "GCF_016906025.1 Haloterrigena salifodinae strain=BOL5-1\n",
      "GCF_011602445.1 Haloarcula sp. JP-L23 strain=JP-L23\n",
      "GCF_018141185.1 Methanobacterium sp. MethCAN strain=MethCAN\n",
      "GCF_013267195.1 Halorubrum salinarum strain=RHB-C\n",
      "GCF_013402875.1 Halobaculum salinum strain=NJ-3-1\n",
      "GCF_013401515.1 Halobaculum halophilum strain=Gai3-2\n",
      "GCF_013391085.1 Natronomonas halophila strain=C90\n",
      "GCF_014876775.1 Infirmifilum lucidum strain=3507LT\n",
      "GCF_017094485.1 Natranaeroarchaeum sulfidigenum strain=AArc-S\n",
      "GCF_009217585.1 Halomicrobium sp. ZPS1 strain=ZPS1\n",
      "GCF_018128925.1 Nitrosopumilus sp. K4 strain=K4\n",
      "GCF_016917835.2 Halobacterium sp. BOL4-2 strain=BOL4-2\n",
      "GCF_016917855.1 Halobacterium sp. GSL-19 strain=GSL-19\n",
      "GCF_017352155.2 Haloterrigena alkaliphila strain=KZCA68\n",
      "GCF_017357405.1 Halovivax sp. KZCA124 strain=KZCA124\n",
      "GCF_021184045.1 Methanococcoides orentis strain=LMO-1\n",
      "GCF_019704315.1 Stygiolobus sp. KN-1 strain=KN-1\n",
      "GCF_018502485.1 Methanospirillum sp. J.3.6.1-F.2.7.3 strain=J.3.6.1-F.2.7.3\n",
      "GCF_018623055.1 Halorubrum sp. 2020YC2 strain=2020YC2\n",
      "GCF_017094445.1 Halapricum desulfuricans strain=HSR-Bgl\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "GCF_017094465.1 Halapricum desulfuricans strain=HSR-Est\n",
      "GCF_017094505.1 Halapricum desulfuricans strain=HSR12-1\n",
      "GCF_017094525.1 Halapricum desulfuricans strain=HSR12-2\n",
      "GCF_019693215.1 Salinarchaeum sp. IM2453 strain=IM2453\n",
      "GCF_023746595.1 Thermococcus sp. IOH2 strain=IOH2\n",
      "GCF_023617305.1 Methanobrevibacter sp. TLL-48-HuF1 strain=TLL-48-HuF1\n",
      "GCF_019879105.1 Natrinema sp. SYSU A 869 strain=SYSU A 869\n",
      "GCF_019880225.1 Halobaculum rubrum strain=C46\n",
      "GCF_020885915.1 Candidatus Methanoliparum sp. LAM-1 strain=LAM-1\n",
      "GCF_020150815.1 Haloprofundus sp. Gai1-5 strain=Gai1-5\n",
      "GCF_020097815.1 Haloprofundus sp. SQT7-1 strain=SQT7-1\n",
      "GCF_020097835.1 Haloprofundus sp. SEDH52 strain=SEDH52\n",
      "GCF_020405185.1 Halomicrobium salinisoli strain=LT50\n",
      "GCF_020405245.1 Halomicrobium salinisoli strain=TH30\n",
      "GCF_020405205.1 Natrinema salinisoli strain=SLN56\n",
      "GCF_020405225.1 Natrinema sp. DC36 strain=DC36\n",
      "GCF_020618475.1 Haladaptatus sp. PSR5 strain=PSR5\n",
      "GCF_020614375.1 Haladaptatus sp. PSR8 strain=PSR8\n",
      "GCF_021655615.1 Acidianus sp. HS-5 strain=HS-5\n",
      "GCF_020736485.1 Halobacterium sp. NRC-34001 strain=NRC-34001\n",
      "GCF_021249345.1 Halobacterium sp. Gai3-17 strain=Gai3-17\n",
      "GCF_021249405.1 Halobacterium sp. XZYJT26 strain=XZYJT26\n",
      "GCF_021431925.1 Halobaculum sp. CBA1158 strain=CBA1158\n",
      "GCF_023028105.1 Halovarius sp. SQT-29-1 strain=SQT-29-1\n",
      "GCF_023028345.1 Halorientalis sp. GDY88 strain=GDY88\n",
      "GCF_023028225.1 Halorientalis sp. ZY14 strain=ZY14\n",
      "GCF_023016325.1 Methanonatronarchaeum sp. AMET6-2 strain=AMET6-2\n",
      "GCF_023115355.1 Halocatena sp. AD-1 strain=AD-1\n",
      "GCF_023238445.1 Halorussus sp. XZYJT40 strain=XZYJT40\n",
      "GCF_023238205.1 Halorussus sp. XZYJT49 strain=XZYJT49\n",
      "GCF_023702125.1 Halorubrum sp. 1-13-X8 strain=1-13-X8\n",
      "GCF_023703775.1 Haloplanus sp. GDY1 strain=GDY1\n",
      "GCF_023703795.1 Haloplanus sp. HW8-1 strain=HW8-1\n",
      "GCF_023913515.1 Natronococcus sp. CG52 strain=CG52\n"
     ]
    }
   ],
   "source": [
    "assembly_summary_path = \"archaea_assembly_summary.txt\"\n",
    "Entrez.email = \"moshenskydenis@gmail.com\"\n",
    "db_dir = \"archaea\"\n",
    "\n",
    "processed_ac = list()\n",
    "arch_df = pd.read_csv(assembly_summary_path, sep=\"\\t\")\n",
    "# arch_df.query(\"assembly_level=='Complete Genome' & refseq_category!='representative genome'\")\n",
    "for _, row in arch_df.query(\"assembly_level=='Complete Genome'\").iterrows():\n",
    "    ac = row['assembly_accession']  # id field in genomes_table\n",
    "    asm = row['asm_name']\n",
    "    taxonomy = get_taxonomy(row['species_taxid'])\n",
    "    name = row['organism_name'] + (\"\" if pd.isna(row['infraspecific_name']) else \" \" + row['infraspecific_name'])\n",
    "    ftp_prefix = row['ftp_path'] + \"/\" + ac + \"_\" + asm\n",
    "    fna_seq = [ftp_prefix+\"_genomic.fna.gz\"]\n",
    "    # fna_seq = [os.path.join(db_dir, ac+\"_genomic.fna\")] #\n",
    "    \n",
    "    try:\n",
    "    #     with open(fna_seq[0], 'wb') as gfna_f: #\n",
    "    #         gfna_f.write(zlib.decompress(urllib.request.urlopen(ftp_prefix+\"_genomic.fna.gz\").read(), 15+32)) #\n",
    "    #     with open(os.path.join(db_dir, ac+\"_rna_from_genomic.fna\"), 'wb') as rna_f:\n",
    "    #         rna_f.write(zlib.decompress(urllib.request.urlopen(ftp_prefix+\"_rna_from_genomic.fna.gz\").read(), 15+32))\n",
    "    #     with open(os.path.join(db_dir, ac+\"_translated_cds.faa\"), 'wb') as tcds_f:\n",
    "    #         tcds_f.write(zlib.decompress(urllib.request.urlopen(ftp_prefix+\"_translated_cds.faa.gz\").read(), 15+32))\n",
    "        processed_ac.append({\"id\": ac, \"name\": name, \"taxonomy\": taxonomy, \"btc_seqs\": [], \"fna_seq\": fna_seq})\n",
    "    except:\n",
    "        print(traceback.format_exc())\n",
    "    print(ac, name)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "254b31e4",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 98,
   "id": "ef4be58c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['GCF_000762265.1', 'GCF_001458655.1', 'GCF_002813085.1']"
      ]
     },
     "execution_count": 98,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "processed_ac"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "c41af9dd",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "# cmbuild :: covariance model construction from multiple sequence alignments\n",
      "# INFERNAL 1.1.4 (Dec 2020)\n",
      "# Copyright (C) 2020 Howard Hughes Medical Institute.\n",
      "# Freely distributed under the BSD open source license.\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "# CM file:                                            16S_rRNA_archaea_ss.cm\n",
      "# alignment file:                                     RF01959.stockholm\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "#                                                                      rel entropy\n",
      "#                                                                      -----------\n",
      "# idx    name                     nseq eff_nseq   alen  clen  bps bifs    CM   HMM description\n",
      "# ------ -------------------- -------- -------- ------ ----- ---- ---- ----- ----- -----------\n",
      "       1 SSU_rRNA_archaea           86     1.06   1958  1478  457   30 0.589 0.302 Archaeal small subunit ribosomal RNA\n",
      "#\n",
      "# CPU time: 16.89u 0.02s 00:00:16.91 Elapsed: 00:00:16.97\n",
      "# cmcalibrate :: fit exponential tails for CM E-values\n",
      "# INFERNAL 1.1.4 (Dec 2020)\n",
      "# Copyright (C) 2020 Howard Hughes Medical Institute.\n",
      "# Freely distributed under the BSD open source license.\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "# CM file:                                     16S_rRNA_archaea_ss.cm\n",
      "# number of worker threads:                    16\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "#\n",
      "# Calibrating CM(s):\n",
      "#\n",
      "#                        predicted                                                   actual   \n",
      "#                       running time              percent complete                running time\n",
      "# model name            (hr:min:sec)  [........25........50........75..........]  (hr:min:sec)\n",
      "# --------------------  ------------  ------------------------------------------  ------------\n",
      "  SSU_rRNA_archaea          00:37:18  ["
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Killed\n"
     ]
    }
   ],
   "source": [
    "arch_16s_profile_ss = SeqsProfile.build(mult_aln=\"RF01959.stockholm\", name=\"16S_rRNA_archaea_ss\", method=\"infernal\", seqs_type=\"rna\", noss=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "04e197f8",
   "metadata": {},
   "outputs": [],
   "source": [
    "arch_16s_profile = SeqsProfile(SeqsProfileInfo.load_from_dict({\n",
    "    'name': '16S_rRNA_archaea',\n",
    "    'path': '16S_rRNA_archaea.cm',\n",
    "    'type': 'rna',\n",
    "    'weight': 1.0,\n",
    "    'method': 'infernal'\n",
    "}))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "4708f6c3",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "# cmsearch :: search CM(s) against a sequence database\n",
      "# INFERNAL 1.1.4 (Dec 2020)\n",
      "# Copyright (C) 2020 Howard Hughes Medical Institute.\n",
      "# Freely distributed under the BSD open source license.\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "# query CM file:                         16S_rRNA_archaea.cm\n",
      "# target sequence database:              archaea/GCF_000762265.1_rna_from_genomic.fna\n",
      "# tabular output of hits:                16S_rRNA_archaea_out_J3o4Z3o0wS.psr\n",
      "# number of worker threads:              4 [--cpu]\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "\n",
      "Query:       RF01959  [CLEN=1478]\n",
      "Hit scores:\n",
      " rank     E-value  score  bias  sequence                   start    end   mdl trunc   gc  description\n",
      " ----   --------- ------ -----  ------------------------- ------ ------   --- ----- ----  -----------\n",
      "  (1) !         0 1293.4  12.7  lcl|NZ_CP006933.1_rrna_4       3   1475 + hmm     - 0.56  [locus_tag=BRM9_RS01355] [db_xref=RFAM:RF01959] [pro\n",
      "  (2) !         0 1293.4  12.7  lcl|NZ_CP006933.1_rrna_25      3   1475 + hmm     - 0.56  [locus_tag=BRM9_RS03745] [db_xref=RFAM:RF01959] [pro\n",
      "\n",
      "\n",
      "Hit alignments:\n",
      ">> lcl|NZ_CP006933.1_rrna_4  [locus_tag=BRM9_RS01355] [db_xref=RFAM:RF01959] [product=16S ribosomal RNA] [location=28338\n",
      " rank     E-value  score  bias mdl mdl from   mdl to       seq from      seq to       acc trunc   gc\n",
      " ----   --------- ------ ----- --- -------- --------    ----------- -----------      ---- ----- ----\n",
      "  (1) !         0 1293.4  12.7 hmm        3     1477 ..           3        1475 + .. 0.99     - 0.56\n",
      "\n",
      "                                ::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::: CS\n",
      "                   RF01959    3 UCCGGUUGAUCCUGCCGGAggccAcuGCUAUcGGggUucGauUaAGCCAUGCgAGUcgagcGuuguuuagggugcguGGCGgA 85  \n",
      "                                UCCG UUGAUCCUG CGGAggccAcuGCUAU+GGg UucGauUaAGCCAUGC AGUcga+cG u++uu ggg+ cguGGCG+A\n",
      "  lcl|NZ_CP006933.1_rrna_4    3 UCCGUUUGAUCCUGGCGGAGGCCACUGCUAUUGGGUUUCGAUUAAGCCAUGCAAGUCGAACGAUCCUUCGGGGUCGUGGCGUA 85  \n",
      "                                7999**********************************************************999999999999********* PP\n",
      "\n",
      "                                ::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::: CS\n",
      "                   RF01959   86 cgGCUcAGUAACACGUGGccAAcCUaCCCUugGGagggGgAUAaCCuCGGGAAACUGgGGcUAAUacccgAUAggugauggau 168 \n",
      "                                cgGCUcAGUAACACGUGG++AAcCUa CCUu GGa+ gGgAUAaCC+ GGGAAACUGgGG+UAAUacc+gAUA gu  +g + \n",
      "  lcl|NZ_CP006933.1_rrna_4   86 CGGCUCAGUAACACGUGGAUAACCUAACCUUAGGACUGGGAUAACCCUGGGAAACUGGGGAUAAUACCGGAUAUGUAGGGCUG 168 \n",
      "                                *********************************************************************************** PP\n",
      "\n",
      "                                ::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::: CS\n",
      "                   RF01959  169 gCUGGAAuGguuugugguugAAAgGuuccggCGCcuuAGGAUGgGgCUGCGGCcgAUuAGGUaGuuGGuGGGGUAAcGGcCCa 251 \n",
      "                                 CUGGAAuGguu+  + uugAAA+Guuccg CGCcu AGGAUGg +CUGCGGC gAUuAGGUaGuuGG GGGGUAA  GcCCa\n",
      "  lcl|NZ_CP006933.1_rrna_4  169 CCUGGAAUGGUUCCCUAUUGAAAUGUUCCGACGCCUAAGGAUGGAUCUGCGGCAGAUUAGGUAGUUGGCGGGGUAAAUGCCCA 251 \n",
      "                                *********************************************************************************** PP\n",
      "\n",
      "                                ::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::: CS\n",
      "                   RF01959  252 cCaaGCCgaugAucGGUAcGGGuugUGaGAGcggGAGCCcgGAGAuGGguuCUGAGACAaGgguCCaGGCCCUACGGGGcGCA 334 \n",
      "                                cCaaGCC  u+Auc GUAcGGGuugUGaGAGc++GAGCCcgGAGAuGG++ CUGAGACAaGg+uCCaGGCCCUACGGGGcGCA\n",
      "  lcl|NZ_CP006933.1_rrna_4  252 CCAAGCCAGUAAUCUGUACGGGUUGUGAGAGCAAGAGCCCGGAGAUGGAACCUGAGACAAGGUUCCAGGCCCUACGGGGCGCA 334 \n",
      "                                *********************************************************************************** PP\n",
      "\n",
      "                                ::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::: CS\n",
      "                   RF01959  335 GCAGGCGCGAAAcCUuugCAaUGCgcGaaAgcGcGAuggGGggAccCCgAGUGccagggcauauagugguggCUuUUcuggag 417 \n",
      "                                GCAGGCGCGAAAcCU++gCAaUGC+cGaaAg+GcGA+ggGGg A cCC AGUGcca + + ua+ g gguggCUuUUcu++ag\n",
      "  lcl|NZ_CP006933.1_rrna_4  335 GCAGGCGCGAAACCUCCGCAAUGCACGAAAGUGCGACGGGGGAAACCCAAGUGCCACUCU-UAACGGGGUGGCUUUUCUUAAG 416 \n",
      "                                ******************************************************999876.56668999************** PP\n",
      "\n",
      "                                ::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::: CS\n",
      "                   RF01959  418 ugUAagaaGguugggGAAuAAGgGcuGGGcAAGaCcGGUGcCAGCCGCCGCGGUAAuACCgGCagCuCGAGUGGUggccgcua 500 \n",
      "                                ugUAa+aaG+uu +gGAAuAAG GcuGGGcAAGaCcGGUGcCAGCCGCCGCGGUAA ACCgGCagCuC AGUGGUggccg+u+\n",
      "  lcl|NZ_CP006933.1_rrna_4  417 UGUAAAAAGCUUUUGGAAUAAGAGCUGGGCAAGACCGGUGCCAGCCGCCGCGGUAACACCGGCAGCUCAAGUGGUGGCCGUUU 499 \n",
      "                                *********************************************************************************** PP\n",
      "\n",
      "                                ::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::: CS\n",
      "                   RF01959  501 UUAUUGGGCCUAAAGcGucCGUAGCcGGccugguaaGUcccuuGggAAAuCcaccgGCUcAACgggugGgcgugcgGggGauA 583 \n",
      "                                UUAUUGGGCCUAAAGcGu+CGUAGCcGGc+ug uaaGUc+cu+G+gAAAuC+  cgGCU+AAC g+  G+ +ugc+Gg GauA\n",
      "  lcl|NZ_CP006933.1_rrna_4  500 UUAUUGGGCCUAAAGCGUUCGUAGCCGGCUUGAUAAGUCUCUGGUGAAAUCUCACGGCUUAACCGUGAGAAUUGCUGGAGAUA 582 \n",
      "                                *********************************************************************************** PP\n",
      "\n",
      "                                ::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::: CS\n",
      "                   RF01959  584 CugcugGgCUuGGGacCGGGaGaggcggggGGUACuuccgGGGUAGGgGUGAAAUCCugUaAUCCugGgaGGACCaCCgGUGG 666 \n",
      "                                Cu +u+GgCUuG G+cCGGGaGagg+ +g GGUACu+ccgGGGUAGGgGUGAAAUCCu UaAUCC+gGgaGGACCaCC+GUGG\n",
      "  lcl|NZ_CP006933.1_rrna_4  583 CUAUUAGGCUUGAGGCCGGGAGAGGUUAGCGGUACUCCCGGGGUAGGGGUGAAAUCCUAUAAUCCCGGGAGGACCACCUGUGG 665 \n",
      "                                *********************************************************************************** PP\n",
      "\n",
      "                                ::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::: CS\n",
      "                   RF01959  667 CGAAgGCgccuggcuaGaACGgguCCGACGGUGAGGGaCGAAaGCugGGGgagCgAAccGGAUUAGAUACCCGGGUAGUCCug 749 \n",
      "                                CGAAgGCg cu +cu GaACGg++C GACGGUGAG  aCGAAaGC++GGGg+gCgAAccGGAUUAGAUACCCGGGUAGUCCug\n",
      "  lcl|NZ_CP006933.1_rrna_4  666 CGAAGGCGGCUAACUGGAACGGACCUGACGGUGAGUAACGAAAGCCAGGGGCGCGAACCGGAUUAGAUACCCGGGUAGUCCUG 748 \n",
      "                                *********************************************************************************** PP\n",
      "\n",
      "                                ::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::: CS\n",
      "                   RF01959  750 GCuGUAAACGaUGcggGCUaGGUGUuggguaggCuucGaGccugugcaGUGcCGuAGgGAAGcCGuuAAGcgcGCCGCCUGGG 832 \n",
      "                                GC+GUAAACGaUG+gg CU+GGUGUuggg +ggCu cGaGc+  ++caGUGcCG AGgGAAGc+GuuAAG++c CCGCCUGGG\n",
      "  lcl|NZ_CP006933.1_rrna_4  749 GCCGUAAACGAUGUGGACUUGGUGUUGGGAUGGCUCCGAGCUGCCCCAGUGCCGAAGGGAAGCUGUUAAGUCCACCGCCUGGG 831 \n",
      "                                *********************************************************************************** PP\n",
      "\n",
      "                                ::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::: CS\n",
      "                   RF01959  833 gAGUACGgcCGCAAGGcUGAAACUUAAAGGAAUUGGCGGGGGAGCACcACAAcgGGuGGAGccUGCGGUUUAAUUGGAuUCAA 915 \n",
      "                                +AGUACGg+CGCAAG cUGAAACUUAAAGGAAUUGGCGGGGGAGCACcACAAcg GuGGAGccUGCGGUUUAAUUGGAuUCAA\n",
      "  lcl|NZ_CP006933.1_rrna_4  832 AAGUACGGUCGCAAGACUGAAACUUAAAGGAAUUGGCGGGGGAGCACCACAACGCGUGGAGCCUGCGGUUUAAUUGGAUUCAA 914 \n",
      "                                *********************************************************************************** PP\n",
      "\n",
      "                                ::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::: CS\n",
      "                   RF01959  916 CGCCGGaaAuCUcACCgGgggcGACaGcaggaUGAaGGuCAGguUGAaGaccUUaCcuGAcuugcuGAGAGGaGGUGCAUGGC 998 \n",
      "                                CGCCGGa+AuCUcACC+GgggcGACaGcag+aUGA  G+CAGguUGA+GaccUU+C+uGAc +gcuGAGAGGaGGUGCAUGGC\n",
      "  lcl|NZ_CP006933.1_rrna_4  915 CGCCGGACAUCUCACCAGGGGCGACAGCAGAAUGAUAGCCAGGUUGAUGACCUUGCUUGACAAGCUGAGAGGAGGUGCAUGGC 997 \n",
      "                                *********************************************************************************** PP\n",
      "\n",
      "                                ::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::: CS\n",
      "                   RF01959  999 CGuCGuCAGcUCGUaCCGUGAGGcGUCCuGUUAAGUCaGGcAACGAGCGAGACCCgcgccccuagUUGCcAgcauguucuucg 1081\n",
      "                                CG+CGuCAGcUCGUaCCGUGAGGcGUCCuGUUAAGUCaGGcAACGAGCGAGACCC cgccc+uagUU CcAgc++ u+cuucg\n",
      "  lcl|NZ_CP006933.1_rrna_4  998 CGCCGUCAGCUCGUACCGUGAGGCGUCCUGUUAAGUCAGGCAACGAGCGAGACCCACGCCCUUAGUUACCAGCGGAUCCUUCG 1080\n",
      "                                *********************************************************************************** PP\n",
      "\n",
      "                                ::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::: CS\n",
      "                   RF01959 1082 GgugGguGgGcACacUaggGGGACuGCCggcGcuAAggcGGAGGAAGGagcGGGCaACGGuAGGUCaGuAUGCCCCGAAucuc 1164\n",
      "                                Gg++G  GgGcACacUa gGGGAC+GCC g+G uAA+ +GGAGGAAGGag GG C ACGGuAGGUC GuAUGCCCCGAAuc+c\n",
      "  lcl|NZ_CP006933.1_rrna_4 1081 GGAUGCCGGGCACACUAAGGGGACCGCCAGUGAUAAACUGGAGGAAGGAGUGGACGACGGUAGGUCCGUAUGCCCCGAAUCCC 1163\n",
      "                                99999****************************************************************************** PP\n",
      "\n",
      "                                ::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::: CS\n",
      "                   RF01959 1165 CuGGGCuACACGCGgGCUaCAAUGGccGgGACAauGGGuugCgACcCCGAaAGGgGgaGguAAUCuCcuAAACccgguCguAG 1247\n",
      "                                CuGGGCuACACGCGgGCUaCAAUGG++ gGACAauGGGuu+CgAC C GAaAGG GgaGguAAUCuCcuAAACc+gg+C+uAG\n",
      "  lcl|NZ_CP006933.1_rrna_4 1164 CUGGGCUACACGCGGGCUACAAUGGUUAGGACAAUGGGUUCCGACACUGAAAGGUGGAGGUAAUCUCCUAAACCUGGCCUUAG 1246\n",
      "                                *********************************************************************************** PP\n",
      "\n",
      "                                ::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::: CS\n",
      "                   RF01959 1248 UucGGAUuGaGGGCUGuAACUCGCCCuCgUGAAgcuGGAaUCcgUAGUAAuCGCgugUCAacAucgcGCGGUGAAUAcGUCCC 1330\n",
      "                                UucGGAUuGaGGGCUGuAACUCGCCCuC+UGAAgcuGGAaU cgUAGUAAuCGCgugUCA +A cgcGCGGUGAAUAcGUCCC\n",
      "  lcl|NZ_CP006933.1_rrna_4 1247 UUCGGAUUGAGGGCUGUAACUCGCCCUCAUGAAGCUGGAAUGCGUAGUAAUCGCGUGUCAUAACCGCGCGGUGAAUACGUCCC 1329\n",
      "                                *********************************************************************************** PP\n",
      "\n",
      "                                ::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::: CS\n",
      "                   RF01959 1331 UGCUCCUUGCACACACCGCCCGUCaaacCAcCCGAGuggGguuugGguGAGGccuuguuuuuuugggggggguCGAaucuggg 1413\n",
      "                                UGCUCCUUGCACACACCGCCCGUCa+ cCAcCC A   gGguuugG+uGAGGcc u  u+uuu gg  ++gguCGAaucu gg\n",
      "  lcl|NZ_CP006933.1_rrna_4 1330 UGCUCCUUGCACACACCGCCCGUCACGCCACCCAAAAAGGGUUUGGAUGAGGCCAUAGUCUUU-GGUUAUGGUCGAAUCUAGG 1411\n",
      "                                *******************************************************99998777.7999*************** PP\n",
      "\n",
      "                                :::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::: CS\n",
      "                   RF01959 1414 cUucGcgAGGGGGguuAAGUCGUAACAAGGUagCCGUAGGGGAAcCUGCGGcUGGAUCACCUCC 1477\n",
      "                                +U+  +gAGG GGg +AAGUCGUAACAAGGUagCCGUAGGGGAAcCUGCGGcUGGAUCACCUCC\n",
      "  lcl|NZ_CP006933.1_rrna_4 1412 UUCUUUGAGGAGGGCGAAGUCGUAACAAGGUAGCCGUAGGGGAACCUGCGGCUGGAUCACCUCC 1475\n",
      "                                **************************************************************98 PP\n",
      "\n",
      ">> lcl|NZ_CP006933.1_rrna_25  [locus_tag=BRM9_RS03745] [db_xref=RFAM:RF01959] [product=16S ribosomal RNA] [location=comp\n",
      " rank     E-value  score  bias mdl mdl from   mdl to       seq from      seq to       acc trunc   gc\n",
      " ----   --------- ------ ----- --- -------- --------    ----------- -----------      ---- ----- ----\n",
      "  (2) !         0 1293.4  12.7 hmm        3     1477 ..           3        1475 + .. 0.99     - 0.56\n",
      "\n",
      "                                 :::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::: CS\n",
      "                    RF01959    3 UCCGGUUGAUCCUGCCGGAggccAcuGCUAUcGGggUucGauUaAGCCAUGCgAGUcgagcGuuguuuagggugcguGGCGg 84  \n",
      "                                 UCCG UUGAUCCUG CGGAggccAcuGCUAU+GGg UucGauUaAGCCAUGC AGUcga+cG u++uu ggg+ cguGGCG+\n",
      "  lcl|NZ_CP006933.1_rrna_25    3 UCCGUUUGAUCCUGGCGGAGGCCACUGCUAUUGGGUUUCGAUUAAGCCAUGCAAGUCGAACGAUCCUUCGGGGUCGUGGCGU 84  \n",
      "                                 7999**********************************************************999999999999******** PP\n",
      "\n",
      "                                 :::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::: CS\n",
      "                    RF01959   85 AcgGCUcAGUAACACGUGGccAAcCUaCCCUugGGagggGgAUAaCCuCGGGAAACUGgGGcUAAUacccgAUAggugaugg 166 \n",
      "                                 AcgGCUcAGUAACACGUGG++AAcCUa CCUu GGa+ gGgAUAaCC+ GGGAAACUGgGG+UAAUacc+gAUA gu  +g \n",
      "  lcl|NZ_CP006933.1_rrna_25   85 ACGGCUCAGUAACACGUGGAUAACCUAACCUUAGGACUGGGAUAACCCUGGGAAACUGGGGAUAAUACCGGAUAUGUAGGGC 166 \n",
      "                                 ********************************************************************************** PP\n",
      "\n",
      "                                 :::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::: CS\n",
      "                    RF01959  167 augCUGGAAuGguuugugguugAAAgGuuccggCGCcuuAGGAUGgGgCUGCGGCcgAUuAGGUaGuuGGuGGGGUAAcGGc 248 \n",
      "                                 +  CUGGAAuGguu+  + uugAAA+Guuccg CGCcu AGGAUGg +CUGCGGC gAUuAGGUaGuuGG GGGGUAA  Gc\n",
      "  lcl|NZ_CP006933.1_rrna_25  167 UGCCUGGAAUGGUUCCCUAUUGAAAUGUUCCGACGCCUAAGGAUGGAUCUGCGGCAGAUUAGGUAGUUGGCGGGGUAAAUGC 248 \n",
      "                                 ********************************************************************************** PP\n",
      "\n",
      "                                 :::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::: CS\n",
      "                    RF01959  249 CCacCaaGCCgaugAucGGUAcGGGuugUGaGAGcggGAGCCcgGAGAuGGguuCUGAGACAaGgguCCaGGCCCUACGGGG 330 \n",
      "                                 CCacCaaGCC  u+Auc GUAcGGGuugUGaGAGc++GAGCCcgGAGAuGG++ CUGAGACAaGg+uCCaGGCCCUACGGGG\n",
      "  lcl|NZ_CP006933.1_rrna_25  249 CCACCAAGCCAGUAAUCUGUACGGGUUGUGAGAGCAAGAGCCCGGAGAUGGAACCUGAGACAAGGUUCCAGGCCCUACGGGG 330 \n",
      "                                 ********************************************************************************** PP\n",
      "\n",
      "                                 :::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::: CS\n",
      "                    RF01959  331 cGCAGCAGGCGCGAAAcCUuugCAaUGCgcGaaAgcGcGAuggGGggAccCCgAGUGccagggcauauagugguggCUuUUc 412 \n",
      "                                 cGCAGCAGGCGCGAAAcCU++gCAaUGC+cGaaAg+GcGA+ggGGg A cCC AGUGcca + + ua+ g gguggCUuUUc\n",
      "  lcl|NZ_CP006933.1_rrna_25  331 CGCAGCAGGCGCGAAACCUCCGCAAUGCACGAAAGUGCGACGGGGGAAACCCAAGUGCCACUCU-UAACGGGGUGGCUUUUC 411 \n",
      "                                 **********************************************************999876.56668999********* PP\n",
      "\n",
      "                                 :::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::: CS\n",
      "                    RF01959  413 uggagugUAagaaGguugggGAAuAAGgGcuGGGcAAGaCcGGUGcCAGCCGCCGCGGUAAuACCgGCagCuCGAGUGGUgg 494 \n",
      "                                 u++agugUAa+aaG+uu +gGAAuAAG GcuGGGcAAGaCcGGUGcCAGCCGCCGCGGUAA ACCgGCagCuC AGUGGUgg\n",
      "  lcl|NZ_CP006933.1_rrna_25  412 UUAAGUGUAAAAAGCUUUUGGAAUAAGAGCUGGGCAAGACCGGUGCCAGCCGCCGCGGUAACACCGGCAGCUCAAGUGGUGG 493 \n",
      "                                 ********************************************************************************** PP\n",
      "\n",
      "                                 :::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::: CS\n",
      "                    RF01959  495 ccgcuaUUAUUGGGCCUAAAGcGucCGUAGCcGGccugguaaGUcccuuGggAAAuCcaccgGCUcAACgggugGgcgugcg 576 \n",
      "                                 ccg+u+UUAUUGGGCCUAAAGcGu+CGUAGCcGGc+ug uaaGUc+cu+G+gAAAuC+  cgGCU+AAC g+  G+ +ugc+\n",
      "  lcl|NZ_CP006933.1_rrna_25  494 CCGUUUUUAUUGGGCCUAAAGCGUUCGUAGCCGGCUUGAUAAGUCUCUGGUGAAAUCUCACGGCUUAACCGUGAGAAUUGCU 575 \n",
      "                                 ********************************************************************************** PP\n",
      "\n",
      "                                 :::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::: CS\n",
      "                    RF01959  577 GggGauACugcugGgCUuGGGacCGGGaGaggcggggGGUACuuccgGGGUAGGgGUGAAAUCCugUaAUCCugGgaGGACC 658 \n",
      "                                 Gg GauACu +u+GgCUuG G+cCGGGaGagg+ +g GGUACu+ccgGGGUAGGgGUGAAAUCCu UaAUCC+gGgaGGACC\n",
      "  lcl|NZ_CP006933.1_rrna_25  576 GGAGAUACUAUUAGGCUUGAGGCCGGGAGAGGUUAGCGGUACUCCCGGGGUAGGGGUGAAAUCCUAUAAUCCCGGGAGGACC 657 \n",
      "                                 ********************************************************************************** PP\n",
      "\n",
      "                                 :::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::: CS\n",
      "                    RF01959  659 aCCgGUGGCGAAgGCgccuggcuaGaACGgguCCGACGGUGAGGGaCGAAaGCugGGGgagCgAAccGGAUUAGAUACCCGG 740 \n",
      "                                 aCC+GUGGCGAAgGCg cu +cu GaACGg++C GACGGUGAG  aCGAAaGC++GGGg+gCgAAccGGAUUAGAUACCCGG\n",
      "  lcl|NZ_CP006933.1_rrna_25  658 ACCUGUGGCGAAGGCGGCUAACUGGAACGGACCUGACGGUGAGUAACGAAAGCCAGGGGCGCGAACCGGAUUAGAUACCCGG 739 \n",
      "                                 ********************************************************************************** PP\n",
      "\n",
      "                                 :::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::: CS\n",
      "                    RF01959  741 GUAGUCCugGCuGUAAACGaUGcggGCUaGGUGUuggguaggCuucGaGccugugcaGUGcCGuAGgGAAGcCGuuAAGcgc 822 \n",
      "                                 GUAGUCCugGC+GUAAACGaUG+gg CU+GGUGUuggg +ggCu cGaGc+  ++caGUGcCG AGgGAAGc+GuuAAG++c\n",
      "  lcl|NZ_CP006933.1_rrna_25  740 GUAGUCCUGGCCGUAAACGAUGUGGACUUGGUGUUGGGAUGGCUCCGAGCUGCCCCAGUGCCGAAGGGAAGCUGUUAAGUCC 821 \n",
      "                                 ********************************************************************************** PP\n",
      "\n",
      "                                 :::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::: CS\n",
      "                    RF01959  823 GCCGCCUGGGgAGUACGgcCGCAAGGcUGAAACUUAAAGGAAUUGGCGGGGGAGCACcACAAcgGGuGGAGccUGCGGUUUA 904 \n",
      "                                  CCGCCUGGG+AGUACGg+CGCAAG cUGAAACUUAAAGGAAUUGGCGGGGGAGCACcACAAcg GuGGAGccUGCGGUUUA\n",
      "  lcl|NZ_CP006933.1_rrna_25  822 ACCGCCUGGGAAGUACGGUCGCAAGACUGAAACUUAAAGGAAUUGGCGGGGGAGCACCACAACGCGUGGAGCCUGCGGUUUA 903 \n",
      "                                 ********************************************************************************** PP\n",
      "\n",
      "                                 :::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::: CS\n",
      "                    RF01959  905 AUUGGAuUCAACGCCGGaaAuCUcACCgGgggcGACaGcaggaUGAaGGuCAGguUGAaGaccUUaCcuGAcuugcuGAGAG 986 \n",
      "                                 AUUGGAuUCAACGCCGGa+AuCUcACC+GgggcGACaGcag+aUGA  G+CAGguUGA+GaccUU+C+uGAc +gcuGAGAG\n",
      "  lcl|NZ_CP006933.1_rrna_25  904 AUUGGAUUCAACGCCGGACAUCUCACCAGGGGCGACAGCAGAAUGAUAGCCAGGUUGAUGACCUUGCUUGACAAGCUGAGAG 985 \n",
      "                                 ********************************************************************************** PP\n",
      "\n",
      "                                 :::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::: CS\n",
      "                    RF01959  987 GaGGUGCAUGGCCGuCGuCAGcUCGUaCCGUGAGGcGUCCuGUUAAGUCaGGcAACGAGCGAGACCCgcgccccuagUUGCc 1068\n",
      "                                 GaGGUGCAUGGCCG+CGuCAGcUCGUaCCGUGAGGcGUCCuGUUAAGUCaGGcAACGAGCGAGACCC cgccc+uagUU Cc\n",
      "  lcl|NZ_CP006933.1_rrna_25  986 GAGGUGCAUGGCCGCCGUCAGCUCGUACCGUGAGGCGUCCUGUUAAGUCAGGCAACGAGCGAGACCCACGCCCUUAGUUACC 1067\n",
      "                                 ********************************************************************************** PP\n",
      "\n",
      "                                 :::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::: CS\n",
      "                    RF01959 1069 AgcauguucuucgGgugGguGgGcACacUaggGGGACuGCCggcGcuAAggcGGAGGAAGGagcGGGCaACGGuAGGUCaGu 1150\n",
      "                                 Agc++ u+cuucgGg++G  GgGcACacUa gGGGAC+GCC g+G uAA+ +GGAGGAAGGag GG C ACGGuAGGUC Gu\n",
      "  lcl|NZ_CP006933.1_rrna_25 1068 AGCGGAUCCUUCGGGAUGCCGGGCACACUAAGGGGACCGCCAGUGAUAAACUGGAGGAAGGAGUGGACGACGGUAGGUCCGU 1149\n",
      "                                 *************99999**************************************************************** PP\n",
      "\n",
      "                                 :::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::: CS\n",
      "                    RF01959 1151 AUGCCCCGAAucucCuGGGCuACACGCGgGCUaCAAUGGccGgGACAauGGGuugCgACcCCGAaAGGgGgaGguAAUCuCc 1232\n",
      "                                 AUGCCCCGAAuc+cCuGGGCuACACGCGgGCUaCAAUGG++ gGACAauGGGuu+CgAC C GAaAGG GgaGguAAUCuCc\n",
      "  lcl|NZ_CP006933.1_rrna_25 1150 AUGCCCCGAAUCCCCUGGGCUACACGCGGGCUACAAUGGUUAGGACAAUGGGUUCCGACACUGAAAGGUGGAGGUAAUCUCC 1231\n",
      "                                 ********************************************************************************** PP\n",
      "\n",
      "                                 :::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::: CS\n",
      "                    RF01959 1233 uAAACccgguCguAGUucGGAUuGaGGGCUGuAACUCGCCCuCgUGAAgcuGGAaUCcgUAGUAAuCGCgugUCAacAucgc 1314\n",
      "                                 uAAACc+gg+C+uAGUucGGAUuGaGGGCUGuAACUCGCCCuC+UGAAgcuGGAaU cgUAGUAAuCGCgugUCA +A cgc\n",
      "  lcl|NZ_CP006933.1_rrna_25 1232 UAAACCUGGCCUUAGUUCGGAUUGAGGGCUGUAACUCGCCCUCAUGAAGCUGGAAUGCGUAGUAAUCGCGUGUCAUAACCGC 1313\n",
      "                                 ********************************************************************************** PP\n",
      "\n",
      "                                 :::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::: CS\n",
      "                    RF01959 1315 GCGGUGAAUAcGUCCCUGCUCCUUGCACACACCGCCCGUCaaacCAcCCGAGuggGguuugGguGAGGccuuguuuuuuugg 1396\n",
      "                                 GCGGUGAAUAcGUCCCUGCUCCUUGCACACACCGCCCGUCa+ cCAcCC A   gGguuugG+uGAGGcc u  u+uuu gg\n",
      "  lcl|NZ_CP006933.1_rrna_25 1314 GCGGUGAAUACGUCCCUGCUCCUUGCACACACCGCCCGUCACGCCACCCAAAAAGGGUUUGGAUGAGGCCAUAGUCUUU-GG 1394\n",
      "                                 ***********************************************************************99998777.79 PP\n",
      "\n",
      "                                 ::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::: CS\n",
      "                    RF01959 1397 gggggguCGAaucugggcUucGcgAGGGGGguuAAGUCGUAACAAGGUagCCGUAGGGGAAcCUGCGGcUGGAUCACCUCC 1477\n",
      "                                   ++gguCGAaucu gg+U+  +gAGG GGg +AAGUCGUAACAAGGUagCCGUAGGGGAAcCUGCGGcUGGAUCACCUCC\n",
      "  lcl|NZ_CP006933.1_rrna_25 1395 UUAUGGUCGAAUCUAGGUUCUUUGAGGAGGGCGAAGUCGUAACAAGGUAGCCGUAGGGGAACCUGCGGCUGGAUCACCUCC 1475\n",
      "                                 99*****************************************************************************98 PP\n",
      "\n",
      "\n",
      "\n",
      "Internal HMM-only pipeline statistics summary: (run for model(s) with zero basepairs)\n",
      "--------------------------------------------------------------------------------------\n",
      "Query model(s):                                          1  (1478 consensus positions)\n",
      "Target sequences:                                       55  (26866 residues searched)\n",
      "Windows passing local HMM SSV      filter:               2  (0.1099); expected (0.02)\n",
      "Windows passing local HMM MSV bias filter:               2  (0.1099); expected (0.02)\n",
      "Windows passing local HMM Viterbi  filter:               2  (0.1099); expected (0.001)\n",
      "Windows passing local HMM Forward  filter:               2  (0.1099); expected (1e-05)\n",
      "Total HMM hits reported:                                 2  (0.1097)\n",
      "\n",
      "# CPU time: 0.09u 0.01s 00:00:00.10 Elapsed: 00:00:00.10\n",
      "//\n",
      "[ok]\n"
     ]
    }
   ],
   "source": [
    "psr = arch_16s_profile.search(seqdb=\"archaea/GCF_000762265.1_rna_from_genomic.fna\", threads=4)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "844f7782",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>target name</th>\n",
       "      <th>accession</th>\n",
       "      <th>query name</th>\n",
       "      <th>accession</th>\n",
       "      <th>mdl</th>\n",
       "      <th>mdl from</th>\n",
       "      <th>mdl to</th>\n",
       "      <th>seq from</th>\n",
       "      <th>seq to</th>\n",
       "      <th>strand</th>\n",
       "      <th>trunc</th>\n",
       "      <th>pass</th>\n",
       "      <th>gc</th>\n",
       "      <th>bias</th>\n",
       "      <th>score</th>\n",
       "      <th>E-value</th>\n",
       "      <th>inc</th>\n",
       "      <th>description of target</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>lcl|NZ_CP006933.1_rrna_4</td>\n",
       "      <td>-</td>\n",
       "      <td>RF01959</td>\n",
       "      <td>-</td>\n",
       "      <td>hmm</td>\n",
       "      <td>3</td>\n",
       "      <td>1477</td>\n",
       "      <td>3</td>\n",
       "      <td>1475</td>\n",
       "      <td>+</td>\n",
       "      <td>-</td>\n",
       "      <td>6</td>\n",
       "      <td>0.56</td>\n",
       "      <td>12.7</td>\n",
       "      <td>1293.4</td>\n",
       "      <td>0</td>\n",
       "      <td>!</td>\n",
       "      <td>[locus_tag=BRM9_RS01355] [db_xref=RFAM:RF01959...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>lcl|NZ_CP006933.1_rrna_25</td>\n",
       "      <td>-</td>\n",
       "      <td>RF01959</td>\n",
       "      <td>-</td>\n",
       "      <td>hmm</td>\n",
       "      <td>3</td>\n",
       "      <td>1477</td>\n",
       "      <td>3</td>\n",
       "      <td>1475</td>\n",
       "      <td>+</td>\n",
       "      <td>-</td>\n",
       "      <td>6</td>\n",
       "      <td>0.56</td>\n",
       "      <td>12.7</td>\n",
       "      <td>1293.4</td>\n",
       "      <td>0</td>\n",
       "      <td>!</td>\n",
       "      <td>[locus_tag=BRM9_RS03745] [db_xref=RFAM:RF01959...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                 target name accession query name accession  mdl mdl from  \\\n",
       "0   lcl|NZ_CP006933.1_rrna_4         -    RF01959         -  hmm        3   \n",
       "1  lcl|NZ_CP006933.1_rrna_25         -    RF01959         -  hmm        3   \n",
       "\n",
       "  mdl to seq from seq to strand trunc pass    gc  bias   score E-value inc  \\\n",
       "0   1477        3   1475      +     -    6  0.56  12.7  1293.4       0   !   \n",
       "1   1477        3   1475      +     -    6  0.56  12.7  1293.4       0   !   \n",
       "\n",
       "                               description of target  \n",
       "0  [locus_tag=BRM9_RS01355] [db_xref=RFAM:RF01959...  \n",
       "1  [locus_tag=BRM9_RS03745] [db_xref=RFAM:RF01959...  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "psr"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "250231ca",
   "metadata": {},
   "outputs": [],
   "source": [
    "hsp70_profile = SeqsProfile(SeqsProfileInfo.load_from_dict({\n",
    "    'name': 'HSP70',\n",
    "    'path': 'HSP70.hmm',\n",
    "    'type': 'protein',\n",
    "    'weight': 1.0,\n",
    "    'method': 'hmmer'\n",
    "}))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "46d4d67e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "# hmmsearch :: search profile(s) against a sequence database\n",
      "# HMMER 3.3.2 (Nov 2020); http://hmmer.org/\n",
      "# Copyright (C) 2020 Howard Hughes Medical Institute.\n",
      "# Freely distributed under the BSD open source license.\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "# query HMM file:                  HSP70.hmm\n",
      "# target sequence database:        archaea/GCF_000762265.1_translated_cds.faa\n",
      "# per-dom hits tabular output:     HSP70_out_iR5bRhCqfP.psr\n",
      "# number of worker threads:        4\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "\n",
      "Query:       HSP70  [M=599]\n",
      "Accession:   PF00012.23\n",
      "Description: Hsp70 protein\n",
      "Scores for complete sequences (score includes all domains):\n",
      "   --- full sequence ---   --- best 1 domain ---    -#dom-\n",
      "    E-value  score  bias    E-value  score  bias    exp  N  Sequence                                   Description\n",
      "    ------- ------ -----    ------- ------ -----   ---- --  --------                                   -----------\n",
      "   1.4e-229  760.4  15.6   2.7e-208  690.1  12.8    2.0  2  lcl|NZ_CP006933.1_prot_WP_048085785.1_2239  [gene=dnaK] [locus_tag=BRM9_RS11\n",
      "    1.2e-14   50.6  10.0    4.6e-13   45.3   8.4    2.4  2  lcl|NZ_CP006933.1_prot_WP_048085701.1_2093  [locus_tag=BRM9_RS10565] [db_xre\n",
      "\n",
      "\n",
      "Domain annotation for each sequence (and alignments):\n",
      ">> lcl|NZ_CP006933.1_prot_WP_048085785.1_2239  [gene=dnaK] [locus_tag=BRM9_RS11300] [db_xref=GeneID:24793409] [protein=m\n",
      "   #    score  bias  c-Evalue  i-Evalue hmmfrom  hmm to    alifrom  ali to    envfrom  env to     acc\n",
      " ---   ------ ----- --------- --------- ------- -------    ------- -------    ------- -------    ----\n",
      "   1 !   73.0   0.1   1.6e-24   1.9e-21       1      74 [.       7      85 ..       7      88 .. 0.93\n",
      "   2 !  690.1  12.8  2.3e-211  2.7e-208     100     599 .]      85     584 ..      82     584 .. 0.99\n",
      "\n",
      "  Alignments for each domain:\n",
      "  == domain 1  score: 73.0 bits;  conditional E-value: 1.6e-24\n",
      "                                                EEEEEEXSSEEEEEEEETTEEEEEXXTTSX....SSEESEEEEXTT.SEEETHHHHHTTTTSGGGEEEX CS\n",
      "                                       HSP70  1 viGiDLGttnsvvavmeggkievianeegn....rttPsvvaFtkk.erlvGeaAkrqavtnpkntvfs 64\n",
      "                                                +iGiDLGt ns+ av+ ggk+++i  +eg+    ++ Ps+vaFt++ +rlvGe+A+rqavtnp++t+++\n",
      "  lcl|NZ_CP006933.1_prot_WP_048085785.1_2239  7 IIGIDLGTSNSAAAVLVGGKPTIIPSAEGAtqygKSFPSYVAFTDDgQRLVGEPARRQAVTNPEKTITA 75\n",
      "                                                69**************************974444789******************************** PP\n",
      "\n",
      "                                                GGGTTTXBTT CS\n",
      "                                       HSP70 65 vKrliGrkfs 74\n",
      "                                                +Kr +G++++\n",
      "  lcl|NZ_CP006933.1_prot_WP_048085785.1_2239 76 IKRQMGTNYK 85\n",
      "                                                *******997 PP\n",
      "\n",
      "  == domain 2  score: 690.1 bits;  conditional E-value: 2.3e-211\n",
      "                                                 EEEETTEEEEHHHHHHHHHHHHHHHHHHHHTSBEEEEEEEEXTTXXHHHHHHHHHHHHHTTXEEEEE CS\n",
      "                                       HSP70 100 eveaagekftpeqisamvLkklketaeaylgkkvtdavitvPaYfndaqRqatkdAgkiAGLnvlRi 166\n",
      "                                                 +v++ g+++tp++isa++L+k+k++aea+lg++v++avitvPaYf+d+qR atkdAg+iAGL+v+R+\n",
      "  lcl|NZ_CP006933.1_prot_WP_048085785.1_2239  85 KVDISGKQYTPQEISAFILQKIKKDAEAFLGEEVKKAVITVPAYFDDNQRTATKDAGTIAGLDVVRL 151\n",
      "                                                 67888************************************************************** PP\n",
      "\n",
      "                                                 EEHHHHHHHHHHCCCSS.SCEEEEEEEEXSSXEEEEEEEEETTEEEEEEEEEETTXSHHHHHHHHHH CS\n",
      "                                       HSP70 167 vNepTAaalaYGldkkd.kernvlvyDlGggtfdvSilelekGvfevkaTagDthlGGedfDealve 232\n",
      "                                                 vNepTAa+laYG+dk++ +e +++v+D+Gggt dv i+e++ Gvfev++T+gDt+lGG d+D+ +++\n",
      "  lcl|NZ_CP006933.1_prot_WP_048085785.1_2239 152 VNEPTAASLAYGIDKANeDELEIMVFDFGGGTLDVTIMEFGGGVFEVRSTSGDTKLGGTDMDASIMN 218\n",
      "                                                 ***************9956779********************************************* PP\n",
      "\n",
      "                                                 HHHHHHHHHHSXXGGGSHHHHHHHHHHHHHHHHHTTTSSCSEEEEEEEEEETTT.EEEEEEEEHHHH CS\n",
      "                                       HSP70 233 hlveefkkkegidlskdkkalqrlkeaaekaKkeLSsskqteinleietvdedg.kdlseklsRakf 298\n",
      "                                                 +l++efkk++g+dl +d++a+qrl+eaaekaK+eLS++ +teinl+++t+++dg k+l+ +l+Rak+\n",
      "  lcl|NZ_CP006933.1_prot_WP_048085785.1_2239 219 YLAAEFKKETGVDLLNDDQAVQRLREAAEKAKIELSTTLKTEINLPFITATQDGpKHLTHTLTRAKL 285\n",
      "                                                 ******************************************************************* PP\n",
      "\n",
      "                                                 HHHHHHHHHHTHHHHHHHHHHHTXXGGGESEEEEESGGGGSHHHHHHHHHHHTSXEBSSSXTTTHHH CS\n",
      "                                       HSP70 299 eelvedllertlepvekalkdaklkkeeidevvlvGGstRiPkvqelvkelfgkepsktvnpDeava 365\n",
      "                                                 eelv+ +++r+  p+e+alkdak++k ++d+++lvGG tR+P vq+ v++++gk+ +++++p e+va\n",
      "  lcl|NZ_CP006933.1_prot_WP_048085785.1_2239 286 EELVDSIIKRCSAPMEQALKDAKMSKGDVDKIILVGGPTRMPIVQKFVEDYIGKPVERGIDPMECVA 352\n",
      "                                                 ******************************************************************* PP\n",
      "\n",
      "                                                 HHHHHHHHHHHSSCCSXSEEEEEBESS-EEEEETTTEEEEEESTTEBSSEEEEEEEEESSTTSSSEE CS\n",
      "                                       HSP70 366 iGAAvqagvlsgeeevkdllllDvtpLslgietlggvmtkliernttiptkksqvfstaadnqtave 432\n",
      "                                                 +GAA+q+gvl+g  e+kdl+llDvtpLslgietlggv+tkliernttiptkksqvfstaadnqt+v+\n",
      "  lcl|NZ_CP006933.1_prot_WP_048085785.1_2239 353 VGAAIQGGVLAG--EIKDLVLLDVTPLSLGIETLGGVFTKLIERNTTIPTKKSQVFSTAADNQTSVD 417\n",
      "                                                 ************..***************************************************** PP\n",
      "\n",
      "                                                 EEEEESSSSBGGGSEEEEEEEEES-SSSSTTSSEEEEEEEE-TTSEEEEEEEETTTCEEEEEEEECC CS\n",
      "                                       HSP70 433 ievlqgerelvkdnkllgsfeldgippaprGvpqieVtfdidanGiltvsakdketgkeeeitieas 499\n",
      "                                                 i+vlqger++++ n++lg+f+l gippaprG+pqieVtfdidanGil+vsakd++tgke+ iti+a \n",
      "  lcl|NZ_CP006933.1_prot_WP_048085785.1_2239 418 IHVLQGERSVASGNTTLGRFQLVGIPPAPRGMPQIEVTFDIDANGILNVSAKDMGTGKEQAITITAP 484\n",
      "                                                 ******************************************************************* PP\n",
      "\n",
      "                                                 SSSBHHHHHHHHHHHHHCHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHCGGGSTHHHHHHHHHHHHH CS\n",
      "                                       HSP70 500 eglseeeiekmvkdaeesaeedkkkkeaieakneaeelvyslekkleeeaekvseaekekvesaiee 566\n",
      "                                                 ++lse+ei++ +++ae++aeedkk++e++e +n+a++++y++ek+l+e a+kv++++k+k+e  ++e\n",
      "  lcl|NZ_CP006933.1_prot_WP_048085785.1_2239 485 NKLSEDEIDQKIHEAEQHAEEDKKRQEEVEIRNNADSMIYTAEKTLDELADKVQPEQKTKIEGLVKE 551\n",
      "                                                 ******************************************************************* PP\n",
      "\n",
      "                                                 HHCHTTSSTHHHHHHHHHHHHHHHHHHHHHHHH CS\n",
      "                                       HSP70 567 Lkeeleeeekeeieakteelkkvvekiaeklye 599\n",
      "                                                 L+e + +++ e+i++kteel+k+v+++++++y+\n",
      "  lcl|NZ_CP006933.1_prot_WP_048085785.1_2239 552 LREVVGGDDLEAIKNKTEELTKAVQEVGAAIYQ 584\n",
      "                                                 ***************************999986 PP\n",
      "\n",
      ">> lcl|NZ_CP006933.1_prot_WP_048085701.1_2093  [locus_tag=BRM9_RS10565] [db_xref=GeneID:24793262] [protein=rod shape-det\n",
      "   #    score  bias  c-Evalue  i-Evalue hmmfrom  hmm to    alifrom  ali to    envfrom  env to     acc\n",
      " ---   ------ ----- --------- --------- ------- -------    ------- -------    ------- -------    ----\n",
      "   1 !    4.4   0.0   0.00094       1.1       2      53 ..      22      70 ..      21      82 .. 0.69\n",
      "   2 !   45.3   8.4   3.9e-16   4.6e-13     131     373 ..     112     342 ..      93     356 .. 0.77\n",
      "\n",
      "  Alignments for each domain:\n",
      "  == domain 1  score: 4.4 bits;  conditional E-value: 0.00094\n",
      "                                                EEEEEXSSEEEEEEEETTEEEEEXXTTSXSSEESEEEEXTTS....EEETHHHHHT CS\n",
      "                                       HSP70  2 iGiDLGttnsvvavmeggkievianeegnrttPsvvaFtkke....rlvGeaAkrq 53\n",
      "                                                +GiDLGt n+vva   g+k +  +        Psvva +k++      vGe+Ak  \n",
      "  lcl|NZ_CP006933.1_prot_WP_048085701.1_2093 22 LGIDLGTLNTVVARPSGDKFDLFK-------IPSVVAVKKEDpgyvLAVGEEAKAM 70\n",
      "                                                7************99888887654.......6777777766422223466666665 PP\n",
      "\n",
      "  == domain 2  score: 45.3 bits;  conditional E-value: 3.9e-16\n",
      "                                                 SBEEEEEEEEXTTXXHHHHHHHHHHHHHTTXEEEEEEEHHHHHHHHHHCCCSSSCEEEEEEEEXSSX CS\n",
      "                                       HSP70 131 kkvtdavitvPaYfndaqRqatkdAgkiAGLnvlRivNepTAaalaYGldkkdkernvlvyDlGggt 197\n",
      "                                                 ++++ +v+ +P    + +++a+ d gk AG n + ++ e  Aaa+  Gl  ++  +  +v+D+G g+\n",
      "  lcl|NZ_CP006933.1_prot_WP_048085701.1_2093 112 ASIDRIVVGIPGDASEVEKKAVEDIGKKAGANYVLVISEGLAAAIGAGLPIAE-ASGTMVIDIGAGS 177\n",
      "                                                 56999***********************************************9.7778899****** PP\n",
      "\n",
      "                                                 EEEEEEEEETTEEEEEEEEEETTXSHHHHHHHHHHHHHHHHHHHHSXXGGGSHHH.HHHHH..HHHH CS\n",
      "                                       HSP70 198 fdvSilelekGvfevkaTagDthlGGedfDealvehlveefkkkegidlskdkka.lqrlk..eaae 261\n",
      "                                                  dv +++l+ G+ ++       + GG+d D+++ve + e ++ + gi  +++ k+ +  ++  ++ e\n",
      "  lcl|NZ_CP006933.1_prot_WP_048085701.1_2093 178 SDVVVISLG-GITDIET----IRSGGDDIDSNIVEKVKELYNVEIGIHEAEKAKIeVGMVHseNDGE 239\n",
      "                                                 *******96.7777764....479***********99999999988854444332122222002222 PP\n",
      "\n",
      "                                                 HHHHHTTTSSCSEEEEEEEEEETTTEEEEEEEEHHHHHHHHHHHHHHTHHHHHHHHHHHTXX..GGG CS\n",
      "                                       HSP70 262 kaKkeLSsskqteinleietvdedgkdlseklsRakfeelvedllertlepvekalkdaklk..kee 326\n",
      "                                                 + K          i  ++et +      ++++  +   + +e+++++++e++ k+l+  + +     \n",
      "  lcl|NZ_CP006933.1_prot_WP_048085701.1_2093 240 NGKTV-------VIGKSMETNKP----QKVEIDSNLVADAAEPIIAKLVEALAKVLERMSPEliSGV 295\n",
      "                                                 33322.......23334444444....4555555555666666666666666666665443330555 PP\n",
      "\n",
      "                                                 ESEEEEESGGGGSHHHHHHHHHHHTSXEBSSSXTTTHHHHHHHHHHH CS\n",
      "                                       HSP70 327 idevvlvGGstRiPkvqelvkelfgkepsktvnpDeavaiGAAvqag 373\n",
      "                                                  ++ v+vGG++ +  ++e++ e +g + + + +p  +va+G A+ a+\n",
      "  lcl|NZ_CP006933.1_prot_WP_048085701.1_2093 296 YNKTVVVGGTSQLKGLKERIYEEVGVPVEISDDPMTVVAKGTAIVAA 342\n",
      "                                                 5789****************************************986 PP\n",
      "\n",
      "\n",
      "\n",
      "Internal pipeline statistics summary:\n",
      "-------------------------------------\n",
      "Query model(s):                            1  (599 nodes)\n",
      "Target sequences:                       2389  (677623 residues searched)\n",
      "Passed MSV filter:                       130  (0.0544161); expected 47.8 (0.02)\n",
      "Passed bias filter:                       60  (0.0251151); expected 47.8 (0.02)\n",
      "Passed Vit filter:                         5  (0.00209293); expected 2.4 (0.001)\n",
      "Passed Fwd filter:                         2  (0.00083717); expected 0.0 (1e-05)\n",
      "Initial search space (Z):               2389  [actual number of targets]\n",
      "Domain search space  (domZ):               2  [number of targets reported over threshold]\n",
      "# CPU time: 0.05u 0.00s 00:00:00.05 Elapsed: 00:00:00.03\n",
      "# Mc/sec: 11656.00\n",
      "//\n",
      "[ok]\n"
     ]
    }
   ],
   "source": [
    "psr1 = hsp70_profile.search(seqdb=\"archaea/GCF_000762265.1_translated_cds.faa\", threads=4)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "feee218c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>target name</th>\n",
       "      <th>accession</th>\n",
       "      <th>tlen</th>\n",
       "      <th>query name</th>\n",
       "      <th>accession</th>\n",
       "      <th>qlen</th>\n",
       "      <th>E-value</th>\n",
       "      <th>seq_score</th>\n",
       "      <th>seq_bias</th>\n",
       "      <th>#</th>\n",
       "      <th>...</th>\n",
       "      <th>domain_score</th>\n",
       "      <th>domain_bias</th>\n",
       "      <th>hmm_from</th>\n",
       "      <th>hmm_to</th>\n",
       "      <th>ali_from</th>\n",
       "      <th>ali_to</th>\n",
       "      <th>env_from</th>\n",
       "      <th>env_to</th>\n",
       "      <th>acc</th>\n",
       "      <th>description of target</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>lcl|NZ_CP006933.1_prot_WP_048085785.1_2239</td>\n",
       "      <td>-</td>\n",
       "      <td>619</td>\n",
       "      <td>HSP70</td>\n",
       "      <td>PF00012.23</td>\n",
       "      <td>599</td>\n",
       "      <td>1.4e-229</td>\n",
       "      <td>760.4</td>\n",
       "      <td>15.6</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>73.0</td>\n",
       "      <td>0.1</td>\n",
       "      <td>1</td>\n",
       "      <td>74</td>\n",
       "      <td>7</td>\n",
       "      <td>85</td>\n",
       "      <td>7</td>\n",
       "      <td>88</td>\n",
       "      <td>0.93</td>\n",
       "      <td>[gene=dnaK] [locus_tag=BRM9_RS11300] [db_xref=...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>lcl|NZ_CP006933.1_prot_WP_048085785.1_2239</td>\n",
       "      <td>-</td>\n",
       "      <td>619</td>\n",
       "      <td>HSP70</td>\n",
       "      <td>PF00012.23</td>\n",
       "      <td>599</td>\n",
       "      <td>1.4e-229</td>\n",
       "      <td>760.4</td>\n",
       "      <td>15.6</td>\n",
       "      <td>2</td>\n",
       "      <td>...</td>\n",
       "      <td>690.1</td>\n",
       "      <td>12.8</td>\n",
       "      <td>100</td>\n",
       "      <td>599</td>\n",
       "      <td>85</td>\n",
       "      <td>584</td>\n",
       "      <td>82</td>\n",
       "      <td>584</td>\n",
       "      <td>0.99</td>\n",
       "      <td>[gene=dnaK] [locus_tag=BRM9_RS11300] [db_xref=...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>lcl|NZ_CP006933.1_prot_WP_048085701.1_2093</td>\n",
       "      <td>-</td>\n",
       "      <td>357</td>\n",
       "      <td>HSP70</td>\n",
       "      <td>PF00012.23</td>\n",
       "      <td>599</td>\n",
       "      <td>1.2e-14</td>\n",
       "      <td>50.6</td>\n",
       "      <td>10.0</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>4.4</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2</td>\n",
       "      <td>53</td>\n",
       "      <td>22</td>\n",
       "      <td>70</td>\n",
       "      <td>21</td>\n",
       "      <td>82</td>\n",
       "      <td>0.69</td>\n",
       "      <td>[locus_tag=BRM9_RS10565] [db_xref=GeneID:24793...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>lcl|NZ_CP006933.1_prot_WP_048085701.1_2093</td>\n",
       "      <td>-</td>\n",
       "      <td>357</td>\n",
       "      <td>HSP70</td>\n",
       "      <td>PF00012.23</td>\n",
       "      <td>599</td>\n",
       "      <td>1.2e-14</td>\n",
       "      <td>50.6</td>\n",
       "      <td>10.0</td>\n",
       "      <td>2</td>\n",
       "      <td>...</td>\n",
       "      <td>45.3</td>\n",
       "      <td>8.4</td>\n",
       "      <td>131</td>\n",
       "      <td>373</td>\n",
       "      <td>112</td>\n",
       "      <td>342</td>\n",
       "      <td>93</td>\n",
       "      <td>356</td>\n",
       "      <td>0.77</td>\n",
       "      <td>[locus_tag=BRM9_RS10565] [db_xref=GeneID:24793...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4 rows × 23 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                  target name accession tlen query name  \\\n",
       "0  lcl|NZ_CP006933.1_prot_WP_048085785.1_2239         -  619      HSP70   \n",
       "1  lcl|NZ_CP006933.1_prot_WP_048085785.1_2239         -  619      HSP70   \n",
       "2  lcl|NZ_CP006933.1_prot_WP_048085701.1_2093         -  357      HSP70   \n",
       "3  lcl|NZ_CP006933.1_prot_WP_048085701.1_2093         -  357      HSP70   \n",
       "\n",
       "    accession qlen   E-value seq_score seq_bias  #  ... domain_score  \\\n",
       "0  PF00012.23  599  1.4e-229     760.4     15.6  1  ...         73.0   \n",
       "1  PF00012.23  599  1.4e-229     760.4     15.6  2  ...        690.1   \n",
       "2  PF00012.23  599   1.2e-14      50.6     10.0  1  ...          4.4   \n",
       "3  PF00012.23  599   1.2e-14      50.6     10.0  2  ...         45.3   \n",
       "\n",
       "  domain_bias hmm_from hmm_to ali_from ali_to env_from env_to   acc  \\\n",
       "0         0.1        1     74        7     85        7     88  0.93   \n",
       "1        12.8      100    599       85    584       82    584  0.99   \n",
       "2         0.0        2     53       22     70       21     82  0.69   \n",
       "3         8.4      131    373      112    342       93    356  0.77   \n",
       "\n",
       "                               description of target  \n",
       "0  [gene=dnaK] [locus_tag=BRM9_RS11300] [db_xref=...  \n",
       "1  [gene=dnaK] [locus_tag=BRM9_RS11300] [db_xref=...  \n",
       "2  [locus_tag=BRM9_RS10565] [db_xref=GeneID:24793...  \n",
       "3  [locus_tag=BRM9_RS10565] [db_xref=GeneID:24793...  \n",
       "\n",
       "[4 rows x 23 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "psr1"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
