{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# imports\n",
    "import os\n",
    "import re\n",
    "import json\n",
    "\n",
    "import pandas as pd\n",
    "from Bio.Seq import Seq\n",
    "\n",
    "from eagle.lib.seqs import SeqsDict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# constants\n",
    "WORK_DIR = \"bacteria\"\n",
    "\n",
    "essential_nucl_fasta = \"bacteria/deg_essential/degseq-p.dat\"\n",
    "essential_annot = \"bacteria/deg_essential/degannotation-p.dat\"\n",
    "\n",
    "nonessential_nucl_fasta = \"bacteria/deg_nonessential/degseq-np.dat\"\n",
    "nonessential_annot = \"bacteria/deg_nonessential/degannotation-np.dat\"\n",
    "\n",
    "chr_id_path = \"bacteria/chr_id.txt\"\n",
    "summary_path = \"bacteria/summary_table.txt\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# lib\n",
    "def get_gtfs(chr_dict, deg_annot_path, deg_seqs_fasta, deg_essential):\n",
    "    annot_df = pd.read_csv(deg_annot_path, sep=\"\\t\")\n",
    "    seqs_dict = SeqsDict.load_from_file(deg_seqs_fasta, low_memory=True)\n",
    "\n",
    "    gtfs_df = pd.DataFrame(list(annot_df.apply(prepare_deg_annot_line, axis=1, args=(chr_dict, seqs_dict, deg_essential))))\n",
    "    summary_list = list()\n",
    "    for deg_org in gtfs_df.groupby(\"org_DEG_ID\"):\n",
    "        summary_list.append(deg_org[1].iloc[0][[\"org_DEG_ID\", \"chr_id\", \"org_name\"]].to_dict())\n",
    "        gtf_df = deg_org[1][pd.notna(deg_org[1][\"seqid\"])][[\"seqid\", \"source\", \"type\", \"start\", \"end\", \"score\", \"strand\", \"frame\", \"attribute\"]]\n",
    "        \n",
    "        if deg_essential:\n",
    "            gtf_path = os.path.join(WORK_DIR, deg_org[0]+\"_essential.gtf\")\n",
    "            summary_list[-1].update({\"essential_genes\": gtf_df.shape[0], \"essential_genes_gtf\": gtf_path})\n",
    "        else:\n",
    "            gtf_path = os.path.join(WORK_DIR, deg_org[0]+\"_nonessential.gtf\")\n",
    "            summary_list[-1].update({\"nonessential_genes\": gtf_df.shape[0], \"nonessential_genes_gtf\": gtf_path})\n",
    "        if not gtf_df.empty:\n",
    "            gtf_df.to_csv(gtf_path.replace(\"DNEG\", \"DEG\"), sep=\"\\t\", index=False, quotechar=\"'\")\n",
    "        gtf_path = None\n",
    "        \n",
    "        print(summary_list[-1])\n",
    "    return pd.DataFrame(summary_list)\n",
    "\n",
    "\n",
    "def prepare_deg_annot_line(row, chr_dict, deg_seqs_dict, deg_essential):\n",
    "    result_dict = {\n",
    "        \"org_DEG_ID\": row[\"#DEG_ORG\"],\n",
    "        \"chr_id\": row[\"#Refseq\"],\n",
    "        \"org_name\": row[\"#Organism\"],\n",
    "        # gtf fields\n",
    "        \"seqid\": None,\n",
    "        \"source\": \"DEG\",\n",
    "        \"type\": \"gene\",\n",
    "        \"start\": int(),\n",
    "        \"end\": int(),\n",
    "        \"score\": \"-\",\n",
    "        \"strand\": None,\n",
    "        \"frame\": \".\",\n",
    "        \"attribute\": json.dumps({\"gene_name\": row[\"#Gene_Name\"]}),\n",
    "    }\n",
    "    if row[\"#Refseq\"] not in chr_dict:\n",
    "        return result_dict    \n",
    "    ori = 1\n",
    "    match = search_in_chr(deg_seqs_dict[row[\"#DEG_AC\"]], chr_dict[row[\"#Refseq\"]])\n",
    "    if match is None:\n",
    "        if deg_essential:\n",
    "            match = search_in_chr(str(Seq(deg_seqs_dict[row[\"#DEG_AC\"]]).reverse_complement()), chr_dict[row[\"#Refseq\"]])\n",
    "        else:\n",
    "            match = search_in_chr(deg_seqs_dict[row[\"#DEG_AC\"]][::-1], chr_dict[row[\"#Refseq\"]])\n",
    "        ori = -1\n",
    "    if match is not None:\n",
    "        result_dict[\"seqid\"] = match[0]\n",
    "        result_dict[\"start\"] = match[1] + 1\n",
    "        result_dict[\"end\"] = match[2]\n",
    "        if ori > 0:\n",
    "            result_dict[\"strand\"] = \"+\"\n",
    "        else:\n",
    "            result_dict[\"strand\"] = \"-\"\n",
    "    else:\n",
    "        print(\"WARNING: no match found for '%s'\" % row[\"#DEG_AC\"])\n",
    "    return result_dict\n",
    "\n",
    "\n",
    "def search_in_chr(seq, chr_seq_dict):\n",
    "    for seq_name in chr_seq_dict:\n",
    "        match = re.search(seq.lower(), chr_seq_dict[seq_name].lower())\n",
    "        if match is not None:\n",
    "            return seq_name.split()[0], match.start(), match.end()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# main\n",
    "chr_dict = dict()\n",
    "chr_id_df = pd.read_csv(chr_id_path, sep=\"\\t\")\n",
    "for i, chr_id in enumerate(chr_id_df[\"chr_id\"]):\n",
    "    chr_dict[chr_id] = SeqsDict.load_from_file(chr_id_df.iloc[i][\"fna_path\"], low_memory=True)\n",
    "\n",
    "essential_summary_df = get_gtfs(chr_dict=chr_dict, deg_annot_path=essential_annot, deg_seqs_fasta=essential_nucl_fasta, deg_essential=True)\n",
    "print(\"INFO: got essential summary\")\n",
    "nonessential_summary_df = get_gtfs(chr_dict=chr_dict, deg_annot_path=nonessential_annot, deg_seqs_fasta=nonessential_nucl_fasta, deg_essential=False)\n",
    "nonessential_summary_df[\"org_DEG_ID\"] = nonessential_summary_df[\"org_DEG_ID\"].apply(lambda org_deg_id: org_deg_id.replace(\"DNEG\", \"DEG\"))\n",
    "\n",
    "summary_df = essential_summary_df.merge(nonessential_summary_df[[\"org_DEG_ID\", \"nonessential_genes\", \"nonessential_genes_gtf\"]], on=\"org_DEG_ID\")\n",
    "summary_df_columns = [\"org_DEG_ID\", \"chr_id\", \"org_name\", \"essential_genes\", \"essential_genes_gtf\", \"nonessential_genes\", \"nonessential_genes_gtf\"]\n",
    "summary_df = summary_df[summary_df_columns].merge(chr_id_df, how='left', on=\"chr_id\")\n",
    "summary_df[summary_df[\"essential_genes\"] > 0].to_csv(os.path.join(WORK_DIR, \"summary_table.txt\"), sep=\"\\t\", index=False)\n",
    "\n",
    "!rm ./.*.dat"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.15rc1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
